---
title: "R Notebook"
output: html_notebook
---

#Load packages
```{r}
library(dplyr)
library(Seurat)
library(tictoc)
library(reticulate)
library(limma)#from Biocmanager
library(ggplot2)
library(viridisLite)
library(patchwork)
library(future)
library(RColorBrewer)
divergentcolors_RYB <- colorRampPalette(rev(brewer.pal(11, "RdYlBu")))#take our 3-color gradient (-, 0, +) and make a gradient of colors
plan('multisession', workers = 5)#cores = 5
options(future.globals.maxSize = 100000 * 1024^2)
```

#Load data
```{r}
# Load the datasets
P21_1.data <- Read10X_h5("/home/pheonix/Documents/Phan2020_P21/P211.h5")
P21_2.data <- Read10X_h5("/home/pheonix/Documents/Phan2020_P21/P212.h5")
P21_3.data <- Read10X_h5("/home/pheonix/Documents/Phan2020_P21/P213.h5")

# Initialize the Seurat object with the raw (non-normalized data).
P21_1 <- CreateSeuratObject(counts = P21_1.data, project = "P21_1", min.cells = 3, min.features = 200)
P21_2 <- CreateSeuratObject(counts = P21_2.data, project = "P21_2", min.cells = 3, min.features = 200)
P21_3 <- CreateSeuratObject(counts = P21_3.data, project = "P21_3", min.cells = 3, min.features = 200)

P21_1
P21_2
P21_3

P21_Mouse <- merge(x = P21_1, y = c(P21_2, P21_3), add.cell.ids = c('P21_1', 'P21_2', 'P21_3'), project = 'Phan2020_Mouse_P21_scRNA')
```

#QC
```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
P21_Mouse[["percent.mt"]] <- PercentageFeatureSet(P21_Mouse, pattern = "^mt-")

# Visualize QC metrics as a violin plot
VlnPlot(P21_Mouse, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(P21_Mouse, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(P21_Mouse, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

#QC Filter
```{r}
P21_Mouse
P21_Mouse <- subset(x = P21_Mouse, 
                          subset = percent.mt < 10)
P21_Mouse

# Visualize QC metrics as a violin plot
VlnPlot(P21_Mouse, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

## SCTransform (Normalize, Scale Data)
```{r, warning=FALSE, message=FALSE}
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
P21_Mouse <- SCTransform(P21_Mouse, vars.to.regress = "percent.mt", verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
P21_Mouse <- RunPCA(P21_Mouse, verbose = FALSE)
ElbowPlot(P21_Mouse, ndims = 50)
```

##UMAP Initial
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
tic('Running UMAP & Making DimPlots')
#run the UMAP function using dimensions informed by elbow plot
P21_Mouse <- RunUMAP(P21_Mouse, dims = 1:30, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
P21_Mouse <- FindNeighbors(P21_Mouse, dims = 1:30, verbose = FALSE)
P21_Mouse <- FindClusters(P21_Mouse, verbose = FALSE, algorithm = 3)#default is algorithm = 1/2 (Louvain), 3 = SLM (scATAC), 4 = Leiden (not installed on server currently - need to pip install leidenalg later)
DimPlot(P21_Mouse, label = TRUE, pt.size = 0.5, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P21_Mouse, label = FALSE, pt.size = 0.5) + NoLegend()#no labels
DimPlot(P21_Mouse, group.by = 'orig.ident', pt.size = 0.5)
#DimPlot(P21_Mouse, label = FALSE)#no numbers on clusters, legend off to side
toc()

#Check QC Metrics on Clustering
FeaturePlot(P21_Mouse, features = c('nFeature_RNA', 'nCount_RNA'))
FeaturePlot(P21_Mouse, features = c('percent.mt'))
```

#FeaturePlot
Igfbp6 was fascia DAG marker (Thompson et al., 2022)
```{r fig.height=7, fig.width=7, message=FALSE, warning=FALSE}
## FeaturePlots: view single-cell gene expression across clusters
GOI <- c('Pdgfra', 'Lum', 'Runx1t1', 'Lef1', 'Sox2', 'Crabp1', 'Prdm1', 'Runx3', 'Prlr', 'a',
         'Mfap5', 'Igfbp6', 'Dpp4', 'Col3a1', 'Fabp4', 'Ly6a',
         'Pdgfrb', 'Vim', 'Lrrc17', 'Rgs5', 'Acta2', 'Lrrc15', 'Itga8', 'Itga6', 'Itgb4',
         'Krt14', 'Krt15', 'Sox9', 'Krt10', 'Krt1', 'Cnfn', 'Krt23', 'Krt8', 'Krt18', 'Krt19',
         'Mki67', 'Coch', 'Alas2', 'Hba-a1', 'Ptprc', 'Cd86', 'Cd3g', 'Ttn',
         'Vegfa', 'Pecam1', 'Cdh5', 'Flt1', 'Kdr', 'Flt4', 'Lyve1', 'Sox10', 'Mbp', 'Mpz', 'Dct', 'Tyrp1', 'Dlk2', 'Dll1', 'Angptl1', 'Angptl2')

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = P21_Mouse,
                    features = GOI[n_plot],
                    pt.size = .75,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()
```

### Change Resolution
```{r message=FALSE, warning=FALSE}
P21_Mouse <- FindClusters(P21_Mouse, verbose = FALSE, algorithm = 3, resolution = 0.5)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(P21_Mouse, label = TRUE, pt.size = 0.5, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P21_Mouse, label = FALSE, pt.size = 0.5) + NoLegend()#no labels (to label in photoshop)
DimPlot(P21_Mouse, label = FALSE, pt.size = 0.5, group.by = 'orig.ident')#no labels (to label in photoshop)
```

Save the R object
```{r}
save(P21_Mouse, file = '~/Desktop/Sean_CellChat/Object/Phan2020_P21_Mouse_AllClusters.RData')#final object w/ labels
```

## Label the Clusters by Cell Type
```{r message=FALSE, warning=FALSE}
#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids.all <- c('0 DP', '1 Coch Fibro', '2 RF', '3 BV', '4 Schwann',
                     '5 LBV', '6 BV', '7 Schwann', '8 PF', '9 Pericyte', 
                     '10 Krtno', '11 BV', '12 Krtno', '13 Melanocyte', '14 Krtno', 
                     '15 Fascia', '16 TCell', '17 DS', '18 APM', '19 Macrophage')
names(new.cluster.ids.all) <- levels(P21_Mouse)
P21_Mouse <- RenameIdents(P21_Mouse, new.cluster.ids.all)
```


## Recolor the UMAP clusters by cell lineage
```{r fig.height=7, fig.width=7}
path_out <- '~/Desktop/Sean_CellChat/UMAP/'
p1 <- DimPlot(P21_Mouse, label = FALSE, pt.size = 1.0, 
              cols = c('#00441b', '#238b45', '#74c476', '#8c6bb1', '#ffab4b',
                     '#88419d', '#7850a7', '#fc8d59', '#006d2c', '#7fcdbb', 
                     '#4d8ad5', '#a58bc2', '#00206a', '#f16913', '#16539e', 
                     '#a1d99b', '#800026', '#41ab5d', '#f768a1', '#a50f15')) + NoLegend()#manual colors w/o legend
p1
ggsave(plot = p1, filename = 'Phan2020 Mouse P21 UMAP (All Clusters).png', path = path_out,
         width = 7, height = 7, units = 'in')
```

## Figure FeaturePlots: view single-cell gene expression across clusters
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
GOI <- c('Angptl4')
path_out <- '~/Desktop/Sean_CellChat/UMAP/FigFeaturePlots/Mouse_P21/'
## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = P21_Mouse,
                    features = GOI[n_plot],
                    pt.size = 0.2,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  ggsave(plot = p1, filename = paste(GOI[n_plot], 'P21 Mouse FeaturePlot.png', sep = ' '), path = path_out,
         width = 5, height = 5, units = 'in')
  n_plot <- n_plot + 1
  }
toc()
```

```{r fig.height=5, fig.width=7}
markers <- c('Itga6', 'Krt10', 'Vegfa', 'Angptl1',  'Angptl2', 'Angptl4', 'Tgfb3', 'Tgfbr2')

hr <- DoHeatmap(P21_Mouse,
                features = markers,
                label = TRUE,
                slot = 'data',
                disp.max = 3) +
  scale_fill_gradientn(colours = divergentcolors_RYB(50))
print(hr)
```

