---
title: "Sean_Cheng2018_Merged_Reanalysis"
output: html_notebook
---

#0. Load packages & data
Requires installation of R.utils package
```{r}
library(dplyr)
library(ggplot2)
library(Seurat)
library(tictoc)
library(reticulate)
library(limma)#from Biocmanager
library(viridisLite)
library(patchwork)
library(data.table)
library(monocle3)#installed via devtools::install_github('cole-trapnell-lab/monocle3')
library(SeuratWrappers)#installed via remotes::install_github('satijalab/seurat-wrappers')
```

#load Cheng et al. 2018 foreskin scRNA data
```{r}
tic("Initializing and Merging Cheng et al. 2018 Foreskin scRNA")

#Load the Cheng et al. foreskin datasets
mat <- fread("~/Documents/Cheng2018/exprMatrix.tsv.gz")
meta <- read.table("~/Documents/Cheng2018/meta.tsv", header=T, sep="\t", as.is=T, row.names=1)
genes = mat[,1][[1]]
genes = gsub(".+[|]", "", genes)
mat = data.frame(mat[,-1], row.names=genes)

#convert tissue and sample metadata columns to factors
head(meta, n=10)#sample & tissue metadata columns are both chr vectors
meta$sample <- as.factor(meta$sample)
head(meta, n=10)#now sample is factor and tissue is chr vector
meta$tissue <- as.factor(meta$tissue)

#filter metadata by sample
sample_ids <- levels(meta$sample)#get vector of sample ids
meta_sample <- meta$sample#extract all ordered

#sample 4 (fore12)
meta_1 <- meta[meta$sample == sample_ids[4], ]#filter the metadata (barcode rows) by sample
mat_1 <- t(mat)#transpose the gene x cell matrix so it is easy to filter the barcodes based on the filtered metadata rownames
mat_1 <- as.data.frame(mat_1)#convert large matrix to dataframe
mat_1name <- rownames(mat_1)#vector of barcodes from raw cell x gene matrix
mat_1name <- gsub(pattern = ".", replacement = "-", x = mat_1name, fixed = TRUE)#find and replace all . in barcode id with - to be consistent with metadata
rownames(mat_1) <- mat_1name#replace . in barcode for cell x matrix with - to have same barcode syntax as meta
mat_1name <- rownames(mat_1)#vector of barcodes from raw cell x gene matrix
meta_1name <- rownames(meta_1)#vector of barcodes from filtered metadata
match_i <- pmatch(x = meta_1name, table = mat_1name)#find index position of individual match of matrix barcode from metadata barcodes
mat_1 <- mat_1[match_i, ]#filter the gene matrix
mat_1 <- as.data.frame(t(mat_1))#transpose back to the gene x cell matrix, then convert to dataframe again
Cheng_4 <- CreateSeuratObject(counts = mat_1, project = 'Cheng_fore12', meta.data=meta_1)#create Seurat object for the isolated sample

#sample 5 (fore8)
meta_1 <- meta[meta$sample == sample_ids[5], ]#filter the metadata (barcode rows) by sample
mat_1 <- t(mat)#transpose the gene x cell matrix so it is easy to filter the barcodes based on the filtered metadata rownames
mat_1 <- as.data.frame(mat_1)#convert large matrix to dataframe
mat_1name <- rownames(mat_1)#vector of barcodes from raw cell x gene matrix
mat_1name <- gsub(pattern = ".", replacement = "-", x = mat_1name, fixed = TRUE)#find and replace all . in barcode id with - to be consistent with metadata
rownames(mat_1) <- mat_1name#replace . in barcode for cell x matrix with - to have same barcode syntax as meta
mat_1name <- rownames(mat_1)#vector of barcodes from raw cell x gene matrix
meta_1name <- rownames(meta_1)#vector of barcodes from filtered metadata
match_i <- pmatch(x = meta_1name, table = mat_1name)#find index position of individual match of matrix barcode from metadata barcodes
mat_1 <- mat_1[match_i, ]#filter the gene matrix
mat_1 <- as.data.frame(t(mat_1))#transpose back to the gene x cell matrix, then convert to dataframe again
Cheng_5 <- CreateSeuratObject(counts = mat_1, project = 'Cheng_fore8', meta.data=meta_1)#create Seurat object for the isolated sample

#sample 6 (fore9)
meta_1 <- meta[meta$sample == sample_ids[6], ]#filter the metadata (barcode rows) by sample
mat_1 <- t(mat)#transpose the gene x cell matrix so it is easy to filter the barcodes based on the filtered metadata rownames
mat_1 <- as.data.frame(mat_1)#convert large matrix to dataframe
mat_1name <- rownames(mat_1)#vector of barcodes from raw cell x gene matrix
mat_1name <- gsub(pattern = ".", replacement = "-", x = mat_1name, fixed = TRUE)#find and replace all . in barcode id with - to be consistent with metadata
rownames(mat_1) <- mat_1name#replace . in barcode for cell x matrix with - to have same barcode syntax as meta
mat_1name <- rownames(mat_1)#vector of barcodes from raw cell x gene matrix
meta_1name <- rownames(meta_1)#vector of barcodes from filtered metadata
match_i <- pmatch(x = meta_1name, table = mat_1name)#find index position of individual match of matrix barcode from metadata barcodes
mat_1 <- mat_1[match_i, ]#filter the gene matrix
mat_1 <- as.data.frame(t(mat_1))#transpose back to the gene x cell matrix, then convert to dataframe again
Cheng_6 <- CreateSeuratObject(counts = mat_1, project = 'Cheng_fore9', meta.data=meta_1)#create Seurat object for the isolated sample


Cheng_4
Cheng_5
Cheng_6

#now create the merged foreskin Seurat object
Cheng_Human <- merge(x = Cheng_4, y = c(Cheng_5, Cheng_6), add.cell.ids = c('C_4', 'C_5', 'C_6'), project = 'Cheng_Human_Merged')

toc()
```

#QC
```{r}
Cheng_Human#view merged seurat object
table(Cheng_Human$orig.ident)#view cell count breakdown between samples within the merged dataset

# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
Cheng_Human[["percent.mt"]] <- PercentageFeatureSet(Cheng_Human, pattern = "^MT")

# Visualize QC metrics as a violin plot
VlnPlot(Cheng_Human, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)
VlnPlot(Cheng_Human, features = c("percent.mt"), ncol = 1)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(Cheng_Human, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Cheng_Human, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```

#QC Filter
```{r}
Cheng_Human <- subset(x = Cheng_Human, 
                          subset = percent.mt < 5 &
                        percent.mt > 1)
Cheng_Human
```

## SCTransform (Normalize, Scale Data)
```{r, warning=FALSE, message=FALSE}
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
Cheng_Human <- SCTransform(Cheng_Human, vars.to.regress = "percent.mt", verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
Cheng_Human <- RunPCA(Cheng_Human, verbose = FALSE)
ElbowPlot(Cheng_Human, ndims = 50)
```

## UMAP (initial)
Pick dimensions to use based on ElbowPlot above
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
tic('Running UMAP & Making DimPlots')
#run the UMAP function using dimensions informed by elbow plot
Cheng_Human <- RunUMAP(Cheng_Human, dims = 1:30, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
Cheng_Human <- FindNeighbors(Cheng_Human, dims = 1:30, verbose = FALSE)
Cheng_Human <- FindClusters(Cheng_Human, verbose = FALSE, algorithm = 3)#default is algorithm = 1/2 (Louvain), 3 = SLM (scATAC), 4 = Leiden (doesn't work on server with large datasets)
DimPlot(Cheng_Human, label = TRUE, pt.size = 0.5, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Cheng_Human, label = FALSE, pt.size = 0.5) + NoLegend()#no labels (to label in photoshop)
DimPlot(Cheng_Human, group.by = 'orig.ident', pt.size = 0.5)
#DimPlot(Cheng_Human, label = FALSE)#no numbers on clusters, legend off to side
toc()

#Check QC Metrics on Clustering
FeaturePlot(Cheng_Human, features = c('nFeature_RNA', 'nCount_RNA'))
FeaturePlot(Cheng_Human, features = c('percent.mt'))
```

## FeaturePlots: view single-cell gene expression across clusters
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
GOI <- c('VIM', 'PECAM1', 'VEGFA', 'SOX9', 'KRT15', "COL7A1",
         'PTPRC', 'CD86', 'CDH1', 'ITGA6', 'KRT14', 'KRT5', 'KRT1', 'KRT10', 'KRT8', 'KRT18',
         'CDSN', 'TOP2A', 'SOX10', 'DCT', 'TYRP1', 'DLL1', 'DLK2', 'LEF1', 'EDAR')

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = Cheng_Human,
                    features = GOI[n_plot],
                    pt.size = 0.1,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
  }
toc()
```

### Change Resolution
```{r message=FALSE, warning=FALSE}
Cheng_Human <- FindClusters(Cheng_Human, verbose = FALSE, algorithm = 3, resolution = 0.5)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(Cheng_Human, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Cheng_Human, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
DimPlot(Cheng_Human, group.by = 'orig.ident', pt.size = 0.5)
```

## Label the Clusters by Cell Type
```{r message=FALSE, warning=FALSE}
#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids.all <- c('0 Diff Krtno', '1 Diff Krtno', '2 Basal Krtno', '3 Basal Krtno', '4 Basal Krtno',
                     '5 Melanocyte', '6 Melanocyte', '7 Div Krtno Cells', '8 Sox9 Krtno', '9 Immune Cell', '10 Immune Cell')
names(new.cluster.ids.all) <- levels(Cheng_Human)
Cheng_Human <- RenameIdents(Cheng_Human, new.cluster.ids.all)
```

Save the R object
```{r}
save(Cheng_Human, file = '~/Desktop/Sean_CellChat/Object/Cheng2018_Fore_AllClusters.RData')#final object w/ labels
```

## Recolor the UMAP clusters by cell lineage
```{r fig.height=7, fig.width=7}
path_out <- '~/Desktop/Sean_CellChat/UMAP/'
p1 <- DimPlot(Cheng_Human, label = FALSE, pt.size = 0.5, 
              cols = c('#7ac9f8', '#64a7e9', '#002783', '#2565b5', '#16539e',
                     '#fc8d59', '#f16913', '#4d8ad5', '#00206a', '#790001', '#cc0000')) + NoLegend()#manual colors w/o legend
p1
ggsave(plot = p1, filename = 'Cheng2018 Human NeoFore UMAP (All Clusters).png', path = path_out,
         width = 7, height = 7, units = 'in')
```

#5. Subset keratinocytes
```{r}
Cheng_krtno <- subset(Cheng_Human, idents = c('0 Diff Krtno', '1 Diff Krtno', '2 Basal Krtno', '3 Basal Krtno', '4 Basal Krtno',
                     '7 Div Krtno Cells'), invert = FALSE)#subset the keratinocyte clusters
Cheng_krtno#view the subset
```

#scTransform
```{r}
tic('Running SCTransform')
Cheng_krtno <- SCTransform(Cheng_krtno, vars.to.regress = "percent.mt", verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
Cheng_krtno <- RunPCA(Cheng_krtno, verbose = FALSE)
ElbowPlot(Cheng_krtno, ndims = 50)
```

Based on ElbowPlot, pick major PCs for next steps
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
Cheng_krtno <- RunUMAP(Cheng_krtno, dims = 1:20, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
Cheng_krtno <- FindNeighbors(Cheng_krtno, dims = 1:20, verbose = FALSE)
Cheng_krtno <- FindClusters(Cheng_krtno, verbose = FALSE, algorithm = 3, resolution = 0.6)#default is algorithm = 1 (Louvain), 3 = SLM
toc()
DimPlot(Cheng_krtno, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Cheng_krtno, label = FALSE, pt.size = 1.0) + NoLegend()#no labels
DimPlot(Cheng_krtno, label = FALSE, pt.size = 1.0, group.by = 'orig.ident')#colored by sample
```

#FeaturePlot
```{r message=FALSE, warning=FALSE}
## FeaturePlots: view single-cell gene expression across clusters
GOI <- c("KRT14", "KRT10", "CNFN", "KRT15", "ITGA6", "ITGB4", 
         "LEF1", "EDAR", "SOX6", "SOX9", "COL17A1", "VEGFA", "TOP2A", "PDGFC", "EDA", "AREG", "EREG", "KRT8", "KRT18", "KRT19")

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = Cheng_krtno,
                    features = GOI[n_plot],
                    pt.size = 0.5,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()
```

### Change Resolution
```{r message=FALSE, warning=FALSE}
Cheng_krtno <- FindClusters(Cheng_krtno, verbose = FALSE, algorithm = 3, resolution = 0.05)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(Cheng_krtno, label = TRUE, pt.size = 0.5, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Cheng_krtno, label = FALSE, pt.size = 0.5) + NoLegend()#no labels (to label in photoshop)
DimPlot(Cheng_krtno, label = FALSE, pt.size = 0.5, group.by = 'orig.ident')#no labels (to label in photoshop)
```

##Figure quality UMAP
```{r fig.height=7, fig.width=7}
path_out <- '~/Desktop/Sean_CellChat/UMAP/'
DimPlot(Cheng_krtno, label = TRUE, pt.size = 1, label.size = 6,
        cols = c('#0c3695', '#f0b627', '#225ea8'))#legend
p1 <- DimPlot(Cheng_krtno, label = FALSE, pt.size = 1,
        cols = c('#0c3695', '#f0b627', '#225ea8')) + NoLegend()#blank
p1

ggsave(plot = p1, filename = 'Cheng2018 Human UMAP (Keratinocytes).png', path = path_out,
         width = 7, height = 7, units = 'in')

```

Save the R object
```{r}
save(Cheng_krtno, file = '/t2tb/Object/Cheng2018_Krtno.RData')
```

#Convert Seurat to CDS and Do Pseudotime in Monocle3
```{r}
#want to transfer Seurat UMAP coordinates to CDS object so pseudotime is on the Seurat clusters
temp1 <- Cheng_krtno@reductions$umap@cell.embeddings#large matrix with rows cell barcodes and 2 columns for the Seurat x/y UMAP coordinates
temp2 <- Cheng_krtno@meta.data$SCT_snn_res.0.05#factor levels for the cluster ids at the resolution used to cluster the Seurat object

cds_chengsub <- as.cell_data_set(Cheng_krtno, group.by = 'SCT_snn_res.0.05')#function from SeuratWrappers to convert to Monocle3 object, graph argument specifies the clustering to transfer (assay to use, and active resolution) to @clusters
#cds_chengsub <- cluster_cells(cds_chengsub, reduction_method = "UMAP", group_cells_by = 'seurat_clusters')#

#Add some info not carried over by SeuratWrappers
cds_chengsub <- estimate_size_factors(cds_chengsub)#add size factor metadata to CDS object
cds_chengsub@rowRanges@elementMetadata@listData[["gene_short_name"]] <- rownames(Cheng_krtno)#correct gene names in CDS object

#Validate UMAP coordinate transfer from Seurat object -> CDS object
temp1 <- Cheng_krtno@reductions$umap@cell.embeddings#large matrix with rows cell barcodes and 2 columns for the Seurat x/y UMAP coordinates
temp2 <- cds_chengsub@int_colData@listData$reducedDims$UMAP#matrix with rows cell barcodes and 2 columns for x/y UMAP coordinates
table(temp1 == temp2)#if all TRUE, seuratwrapper successfully transferred umap coordinates to CDS object via the graph argument
#cds_chengsub@int_colData@listData$reducedDims$UMAP <- Cheng_krtno@reductions$umap@cell.embeddings#not needed due to graph argument when making CDS object

#need to transfer Seurat PCA since this wasn't transferred by seurat wrapper
reducedDim(cds_chengsub, type = "PCA") <- Cheng_krtno@reductions$pca@cell.embeddings#transfer the matrix results of Seurat PCA to CDS
#cds_chengsub@preprocess_aux$prop_var_expl <- Cheng_krtno@reductions$pca@stdev#transfer numerical vector containing stdev values per PCA dimension

#partitions still need to be created for learn_graph()
manual_partition <- as.vector(Cheng_krtno@meta.data$seurat_clusters)#get cluster id #s to construct manual partition from a factor
names(manual_partition) <- rownames(Cheng_krtno@meta.data)#
manual_partition <- as.factor(manual_partition)
head(manual_partition)#view ordered levels
levels(manual_partition) <- c(1, 1, 1)#replace cluster id levels with new ids for partitioning in Monocle3 (as.factor ordered 0, 1, 10, 2, 3, etc)
table(levels(manual_partition))#validate replacement
cds_chengsub@clusters@listData$UMAP$partitions <- manual_partition

cds_chengsub@clusters@listData$UMAP$orig.ident <- Cheng_krtno$orig.ident#transfer sample ids to cds object

#recreate.partition <- c(rep(1, length(cds_chengsub@colData@rownames)))
#names(recreate.partition) <- cds_chengsub@colData@rownames#assigns partition value to cell barcode
#recreate.partition <- as.factor(recreate.partition)#converts this to a large factor with 1 level
#cds_chengsub@clusters@listData$UMAP$partitions <- recreate.partition

p1 <- plot_cells(cds_chengsub, reduction_method = 'UMAP',
                 color_cells_by = 'seurat_clusters', 
                 label_cell_groups = FALSE,
                 cell_size = 1,
                 show_trajectory_graph = FALSE) +
  scale_color_manual(values = c('#0c3695', '#f0b627', '#225ea8'))#is a ggplot2 object, so can custom color same as with Seurat's DimPlot
p1
plot_cells(cds_chengsub, reduction_method = 'UMAP',
                 color_cells_by = 'cluster', 
                 label_cell_groups = FALSE,
                 cell_size = 1,
                 show_trajectory_graph = FALSE) +
  scale_color_manual(values = c('#0c3695', '#f0b627', '#225ea8'))#Monocle3 clusters, transferred from Seurat
plot_cells(cds_chengsub, reduction_method = 'UMAP',
                 color_cells_by = 'partition', 
                 label_cell_groups = TRUE,
                 cell_size = 1,
                 show_trajectory_graph = FALSE)
#pseudotime

#rownames(cds_chengsub@principal_graph_aux$UMAP$dp_mst) <- NULL
#colnames(cds_chengsub@int_colData@listData$reducedDims$UMAP) <- NULL

cds_chengsub <- learn_graph(cds_chengsub, use_partition = TRUE, verbose = TRUE)
p2 <- plot_cells(cds_chengsub, reduction_method = 'UMAP',
                 color_cells_by = 'seurat_clusters', 
                 group_label_size = 5, cell_size = 1,
                 show_trajectory_graph = TRUE) +
  scale_color_manual(values = c('#0c3695', '#f0b627', '#225ea8'))#is a ggplot2 object, so can custom color same as with Seurat's DimPlot
p2

plot_cells(cds_chengsub, reduction_method = 'UMAP',
                 color_cells_by = 'partition', 
                 group_label_size = 5, cell_size = 1,
                 show_trajectory_graph = TRUE)#is a ggplot2 object, so can custom color same as with Seurat's DimPlot


cds_chengsub <- order_cells(cds_chengsub)#choose root node
plot_cells(cds_chengsub,
           color_cells_by = "pseudotime",
           group_cells_by = "cluster",
           label_cell_groups = FALSE,
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           label_roots = FALSE,
           cell_size = 1,
           trajectory_graph_color = "grey60")
plot_cells(cds_chengsub,
           color_cells_by = "pseudotime",
           group_cells_by = "cluster",
           label_cell_groups = FALSE,
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           label_roots = FALSE,
           cell_size = 1,
           show_trajectory_graph = FALSE)
```

Save CDS object
```{r}
save(cds_chengsub, file = '/t2tb/Object/Cheng2018_krtno_CDS.RData')
```

#View Expression Trajectories Across Keratinocyte Clusters (w/ GAM)
```{r}
pseudotime_df <- as.data.frame(t(cds_chengsub@assays@data@listData$logcounts))#get cell by gene expression matrix
pseudotime_df$Barcode <- rownames(pseudotime_df)#get column of barcode values
pseudotime_df$Pseudotime <- pseudotime(cds_chengsub)#get pseudotime values

temp_df <- data.frame("Barcode" = rownames(as.data.frame(cds_chengsub@clusters@listData$UMAP$clusters)), 
                      "Cluster" = cds_chengsub@clusters@listData$UMAP$clusters)#get cluster ids

pseudotime_df <- merge(temp_df, pseudotime_df)#filter to shared rows in subset cds
pseudotime_df <- pseudotime_df[pseudotime_df$Pseudotime != Inf, ]#remove rows w/ inf pseudotime value

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey40', se = FALSE) +
  stat_smooth(aes(y = PTCH1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue', se = FALSE) +
  stat_smooth(aes(y = PTCH2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +  
  #stat_smooth(aes(y = SHH), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +   
  stat_smooth(aes(y = WNT10B), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red3', se = FALSE) +
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red4', se = FALSE) +
  stat_smooth(aes(y = EDA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid1', se = FALSE) +   
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal Bud DEGs")#SHH not detected in this dataset

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey40', se = FALSE) +
  #stat_smooth(aes(y = SHH), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +   
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red4', se = FALSE) +
  #stat_smooth(aes(y = EDA), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'orchid1', se = FALSE) +   
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal Bud")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey60', se = FALSE, linetype = 'dashed') +
  #stat_smooth(aes(y = SHH), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'blue', se = FALSE) +   
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red', se = FALSE) +
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'green', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal Bud (Figure)")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = KRT15), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +
  stat_smooth(aes(y = KRT14), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue3', se = FALSE) +  
  stat_smooth(aes(y = KRT5), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +
  stat_smooth(aes(y = KRT10), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod2', se = FALSE) +
  stat_smooth(aes(y = KRT1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkorange', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Keratins")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey30', se = FALSE) +
  stat_smooth(aes(y = ITGA2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey50', se = FALSE) +
  stat_smooth(aes(y = SOX6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +
  stat_smooth(aes(y = SOX9), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'peru', se = FALSE) +  
  stat_smooth(aes(y = DLK2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue3', se = FALSE) +
  stat_smooth(aes(y = NRG1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +
  stat_smooth(aes(y = PHLDB2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  stat_smooth(aes(y = LAMB3), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid4', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal IFE DEGs")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = VEGFA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid1', se = FALSE) +
  #stat_smooth(aes(y = VEGFC), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  #stat_smooth(aes(y = ANGPTL1), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'mediumpurple1', se = FALSE) +
  stat_smooth(aes(y = ANGPTL2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple3', se = FALSE) +
  stat_smooth(aes(y = ANGPTL4), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple4', se = FALSE) +
  stat_smooth(aes(y = PDGFA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgreen', se = FALSE) +
  stat_smooth(aes(y = PDGFB), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'lightseagreen', se = FALSE) +  
  stat_smooth(aes(y = PDGFC), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'green3', se = FALSE) +
  #stat_smooth(aes(y = COCH), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'darkgreen', se = FALSE) +
  #stat_smooth(aes(y = COL2A1), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'lightseagreen', se = FALSE) +  
  theme_classic() +
  ggtitle("Pseudotime Dermal Pocket Genes")


ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = RDH10), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid1', se = FALSE) +
  stat_smooth(aes(y = DHRS3), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkred', se = FALSE) +
  stat_smooth(aes(y = ALDH1A1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple3', se = FALSE) +
  stat_smooth(aes(y = ALDH1A3), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple4', se = FALSE) +
  stat_smooth(aes(y = RARA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +
  stat_smooth(aes(y = RARG), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue3', se = FALSE) +
  stat_smooth(aes(y = CRABP2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime RA Synthesis Genes")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  #stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'grey40', se = FALSE) +
  stat_smooth(aes(y = TGFA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue', se = FALSE) +
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red4', se = FALSE) +
  stat_smooth(aes(y = BMP2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod1', se = FALSE) +
  stat_smooth(aes(y = BMP7), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod3', se = FALSE) +  
  stat_smooth(aes(y = SMAD7), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod', se = FALSE) +
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid', se = FALSE) +
  theme_classic() +
  ggtitle("Glover2023 Regulators of Fingerprint Ridge")
```


```{r}
plot_cells(cds_chengsub,
           color_cells_by = "pseudotime",
           group_cells_by = "cluster",
           label_cell_groups = FALSE,
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           label_roots = TRUE,
           cell_size = 0.75,
           trajectory_graph_color = "grey40", show_trajectory_graph = TRUE)#view pseudotime UMAP with root node
```
