---
title: "R Notebook"
output: html_notebook
---

#Load packages
```{r}
library(dplyr)
library(Seurat)
library(tictoc)
library(reticulate)
library(limma)#from Biocmanager
library(viridisLite)
library(RColorBrewer)
library(ggplot2)
library(patchwork)
library(future)
plan('multisession', workers = 5)#cores = 5
options(future.globals.maxSize = 100000 * 1024^2)

```


#1. Load data & create Seurat object
```{r}
# Load the e90 dataset
e90.data <- Read10X_h5("/home/pheonix/Documents/h5files/e90pig.h5")

# Initialize the Seurat object with the raw (non-normalized data).
e90 <- CreateSeuratObject(counts = e90.data, project = "E90", min.cells = 3, min.features = 200)

e90
```

##QC
```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
e90[["percent.mt"]] <- PercentageFeatureSet(e90, pattern = "^MT")

# Visualize QC metrics as a violin plot
VlnPlot(e90, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(e90, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(e90, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```

#QC filtering
```{r}
e90#before QC filter

e90 <- subset(e90, subset = nCount_RNA > 1900 & nCount_RNA < 25000 & percent.mt < 5)

e90
```

## SCTransform (Normalize, Scale Data)
```{r, warning=FALSE, message=FALSE}
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
e90 <- SCTransform(e90, vars.to.regress = "percent.mt", verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
e90 <- RunPCA(e90, verbose = FALSE)
ElbowPlot(e90, ndims = 50)
toc()
```

## UMAP (initial)
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
e90 <- RunUMAP(e90, dims = 1:40, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
e90 <- FindNeighbors(e90, dims = 1:40, verbose = FALSE)
e90 <- FindClusters(e90, verbose = FALSE, algorithm = 3)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(e90, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(e90, label = FALSE, pt.size = 1.0) + NoLegend()
#DimPlot(e90, label = FALSE)#no numbers on clusters, legend off to side

toc()
```


## FeaturePlots: view single-cell gene expression across clusters
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
GOI <- c('nCount_RNA',
         'PDGFRA', 'LRRC17', 'RUNX1T1', 'VIM', 'ACTA2', 'MYL1', 'MYL2', 
         'LEF1', 'CRABP1', 'RGS5', 'PECAM1', 'CDH5', 'CDH11', 'LYVE1', 'FLT4', 'PTPRC', 'CD86', 'CDH1', 'KRT14', 'KRT5', 'KRT1', 'KRT10', 'KRT8', 'SOX6', 'SOX9',
         'TOP2A', 'CEBPA', 'PPARG', 'FABP4', 'ADIPOQ', 'BMP7', 'VEGFA', 'SOX10', 'COL11A1')
#path_out <- '.'
## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = e90,
                    features = GOI[n_plot],
                    pt.size = 0.5,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
  }
toc()
```

### Change Resolution
```{r message=FALSE, warning=FALSE}
e90 <- FindClusters(e90, verbose = FALSE, algorithm = 3, resolution = 0.4)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(e90, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(e90, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

## Differential Gene Expression
```{r}
tic('ClusterMarker auto-export loop')#start timer
path_out <- '~/Desktop/Sean_CellChat/DEG/E90/'#path to export folder
COI <- levels(e90)#define the clusters you wish to know diff. gene expression of

n_genes <- length(e90@assays[["SCT"]]@data@Dimnames[[1]])#get # of genes we will be testing
n_loops <- 1
while (n_loops < length(COI) + 1) {
#Find positive markers (upregulated genes)
cluster.markers.temp <- FindMarkers(e90, ident.1 = COI[n_loops], min.pct = 0.05, logfc.threshold = 0.0, assay = 'SCT', only.pos = TRUE)#iterate through the entire COI list
#head(cluster.markers.temp, n = 30)
cluster.markers.temp$Gene = rownames(cluster.markers.temp)#add a column for the gene name
cluster.markers.temp$pct.2[cluster.markers.temp$pct.2 == 0] <- NA#corrects pct ratio error if pct.2=0
cluster.markers.temp <- na.omit(cluster.markers.temp)#remove rows with NA
cluster.markers.temp$Pct_Ratio = cluster.markers.temp$pct.1 / cluster.markers.temp$pct.2#compute ratio of how much the cluster of interest expresses compared to all other cells
cluster.markers.temp <- cluster.markers.temp %>% dplyr::select('Gene', everything())#move gene column to the front of the dataframe
cluster.markers.temp$'PctRatio_x_logfc' = cluster.markers.temp$Pct_Ratio * cluster.markers.temp$avg_log2FC
cluster.markers.temp <- cluster.markers.temp[cluster.markers.temp$PctRatio_x_logfc > quantile(cluster.markers.temp$PctRatio_x_logfc, 0.95), ]#keep only the top 95% of upregulated genes in ident.1
cluster.markers.temp <- cluster.markers.temp[order(cluster.markers.temp$PctRatio_x_logfc, decreasing = TRUE), ]#rank genes
write.csv(cluster.markers.temp, file = paste(path_out, (n_loops-1), '_pos_markers.csv', sep = ''), row.names = FALSE)#save that cluster's markers

n_loops <- n_loops + 1#track for when all clusters of one dataset have been run through
}
toc()#end timer
```
24: KRT8/18 (Merkel Cell)
15, 18 is Muscle (Myl1/2)
Cluster 24 DEGs KRT20, 18, and 8 marks Merkel Cells (https://pubmed.ncbi.nlm.nih.gov/7769257/)

## Figure FeaturePlots: view single-cell gene expression across clusters
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
GOI <- c('ANGPTL4')
path_out <- '~/Desktop/Sean_CellChat/UMAP/FigFeaturePlots/Pig_E90/'
## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = e90,
                    features = GOI[n_plot],
                    pt.size = 0.2,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  ggsave(plot = p1, filename = paste(GOI[n_plot], 'E90 Pig FeaturePlot.png', sep = ' '), path = path_out,
         width = 5, height = 5, units = 'in')
  n_plot <- n_plot + 1
  }
toc()
```

##View Top DEGs
```{r fig.height=20, fig.width=14}
path_in <- '~/Desktop/Sean_CellChat/DEG/E90/'#path to import folder

DimPlot(e90, label = TRUE, pt.size = 3, label.size = 9.0)

#load DEGs
temp_clusters <- levels(e90$seurat_clusters)
n_cluster <- 0
while (n_cluster < length(temp_clusters)) {
  temp.csv <- read.csv(file = paste(path_in, n_cluster, '_pos_markers.csv', sep = ''))#read in csv file of DEGs
  temp.csv <- temp.csv[temp.csv$pct.1 >= 0.5, ]#filter to just DEGs expressed in greater than 50% of the cells in the cluster
  temp.csv <- temp.csv[1:10, ]#keep top 10 DEGs
  temp_degs <- temp.csv$Gene#convert top 10 DEGs to a string vector
  
  if (n_cluster == 0) {
    degs <- temp_degs#start storage variable w/ first cluster's top 10 degs
  }
  else {
    degs <- c(degs, temp_degs)#append the cluster's top 10 degs to the storage variable
  }
  
  n_cluster <- n_cluster + 1#iterate to next cluster
}

E90degs <- DoHeatmap(e90,
                features = degs,
                label = TRUE,
                slot = 'data',
                disp.max = 4) +
  scale_fill_gradientn(colours = magma(50))
print(E90degs)

E90degs <- DoHeatmap(e90,
                features = c('PDGFRA', 'ITGA6', 'KRT10', 'EDAR', 'LEF1', 'RGS5', 'PTPRC', 'ACTA2', 'PPARG'),
                label = TRUE,
                slot = 'data',
                disp.max = 4) +
  scale_fill_gradientn(colours = magma(50))
print(E90degs)
```


## Label the Clusters by Cell Type
```{r message=FALSE, warning=FALSE}
#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids <- c('0 K14 Krtno', '1 K10 Krtno', '2 RF', '3 K14 Krtno', '4 DP', 
                     '5 mSweat', '6 PF', '7 K14 Krtno', 
                     '8 weird Fibro', '9 BV', '10 Fibro',
                     '11 mSweat', '12 Pericyte', '13 Macrophage',
                     '14 K10 Krtno', '15 Muscle', '16 K14 Krtno',
                     '17 Adipocyte', '18 Muscle', '19 Schwann Cell', '20 Melanocyte',
                     '21 Immune Cell', '22 Krt14 Krtno', '23 Lymphatic?', '24 Nerve')
names(new.cluster.ids) <- levels(e90)
e90 <- RenameIdents(e90, new.cluster.ids)
```

Save the Seurat object
```{r}
save(e90, file = '~/Desktop/Sean_CellChat/Object/e90_pig_allclusters.RData')
```

## Recolor the UMAP clusters by cell lineage
```{r fig.height=7, fig.width=7}
path_out <- '~/Desktop/Sean_CellChat/UMAP/'
p1 <- DimPlot(e90, label = FALSE, pt.size = 1, 
              cols = c('#002783', '#64a7e9', '#41ab5d', '#16539e', '#00441b', 
                     '#0c6395', '#006837', '#2565b5', '#a1d99b', '#8c6bb1', 
                     '#238b45', '#1179a3', '#7fcdbb', '#790001', '#96ccff', 
                     '#f768a1', '#00206a', '#bdbdbd', '#ae017e', '#ffab4b', 
                     '#f16913', '#BB0000', '#164c9a', '#88419d', '#fc8d59')) + NoLegend()#manual colors w/o legend
p1
ggsave(plot = p1, filename = 'Pig E90 UMAP (All Clusters).png', path = path_out,
         width = 7, height = 7, units = 'in')
```


#2. Subset the Keratinocytes
Subsetting the entire keratinocyte supercluster
```{r message=FALSE, warning=FALSE}
E90_krtno <- subset(e90, idents = c('0 K14 Krtno', '1 K10 Krtno', '3 K14 Krtno', '5 mSweat', '7 K14 Krtno', 
                     '11 mSweat', '14 K10 Krtno', '16 K14 Krtno', '18 Muscle', '20 Melanocyte', '22 Krt14 Krtno'), invert = FALSE)#subset the keratinocyte supercluster
E90_krtno#view the subset

## We want to renormalize the data to bring out the heterogenetiy within the lineage
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
E90_krtno <- SCTransform(E90_krtno, vars.to.regress = "percent.mt", verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
E90_krtno <- RunPCA(E90_krtno, verbose = FALSE)
ElbowPlot(E90_krtno, ndims = 50)
```

Based on ElbowPlot, pick major PCs for next steps
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
E90_krtno <- RunUMAP(E90_krtno, dims = 1:30, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
E90_krtno <- FindNeighbors(E90_krtno, dims = 1:30, verbose = FALSE)
E90_krtno <- FindClusters(E90_krtno, verbose = FALSE, algorithm = 3, resolution = 0.6)#default is algorithm = 1 (Louvain), 3 = SLM
toc()
DimPlot(E90_krtno, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(E90_krtno, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

### View Gene Expression of Keratinocyte Heterogeneity Markers
Includes some markers identified by Sole-Boldo et al. in their analysis as well as DAGs from Thompson et al. 2022
```{r fig.height=5, fig.width=7}
## FeaturePlots: view single-cell gene expression across clusters
DimPlot(E90_krtno, label = TRUE, pt.size = 2, label.size = 6.0)

GOI <- c('PDGFRA', 'APOE', 'KRT14', 'KRT5', 'KRT10', 'KRT18', 'KRT8', 'CALML5', 'CDSN', 'CNFN', 'CDH1',
         'TOP2A', 'ASPM', 'ITGA6', 'ITGB1', 'SOX6', 'SOX9', 'COL17A1', 'DLL1', 'KRT15', 'VEGFA', 'KRT10', 'nFeature_SCT', 'TGFBI', 'DPP4', 'ACTA2', 'PPARG', 'LEF1', 'CTNNB1')

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = E90_krtno,
                    features = GOI[n_plot],
                    pt.size = 1.0,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()

```

### Change Resolution
```{r message=FALSE, warning=FALSE}
E90_krtno <- FindClusters(E90_krtno, verbose = FALSE, algorithm = 3, resolution = 0.5)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(E90_krtno, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(E90_krtno, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

# Differential Gene Expression Within Keratinocytes
```{r}
tic('ClusterMarker auto-export loop')#start timer
path_out <- '~/Desktop/Sean_CellChat/DEG/Krtno_sub/E90/'#path to export folder
COI <- c('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12')#define the clusters you wish to know diff. gene expression of

n_genes <- length(E90_krtno@assays[["SCT"]]@data@Dimnames[[1]])#get # of genes we will be testing
n_loops <- 1
while (n_loops < length(COI) + 1) {
#Find positive markers (upregulated genes)
cluster.markers.temp <- FindMarkers(E90_krtno, ident.1 = COI[n_loops], min.pct = 0.05, logfc.threshold = 0.0, assay = 'SCT', only.pos = TRUE)#iterate through the entire COI list
#head(cluster.markers.temp, n = 30)
cluster.markers.temp$Gene = rownames(cluster.markers.temp)#add a column for the gene name
cluster.markers.temp$pct.2[cluster.markers.temp$pct.2 == 0] <- NA#corrects pct ratio error if pct.2=0
cluster.markers.temp <- na.omit(cluster.markers.temp)#remove rows with NA
cluster.markers.temp$Pct_Ratio = cluster.markers.temp$pct.1 / cluster.markers.temp$pct.2#compute ratio of how much the cluster of interest expresses compared to all other cells
cluster.markers.temp <- cluster.markers.temp %>% dplyr::select('Gene', everything())#move gene column to the front of the dataframe
cluster.markers.temp$'PctRatio_x_logfc' = cluster.markers.temp$Pct_Ratio * cluster.markers.temp$avg_log2FC
cluster.markers.temp <- cluster.markers.temp[cluster.markers.temp$PctRatio_x_logfc > quantile(cluster.markers.temp$PctRatio_x_logfc, 0.95), ]#keep only the top 95% of upregulated genes in ident.1
cluster.markers.temp <- cluster.markers.temp[order(cluster.markers.temp$PctRatio_x_logfc, decreasing = TRUE), ]#rank genes
write.csv(cluster.markers.temp, file = paste(path_out, COI[n_loops], '_pos_markers.csv', sep = ''), row.names = FALSE)#save that cluster's markers

n_loops <- n_loops + 1#track for when all clusters of one dataset have been run through
}
toc()#end timer
```
0 Basal Krtno : DLK2, SOX6, KRT14
1 Diff Krtno : ENSSSCG00000041481 (DMKN), CSTA, SERPINB2, CALML5, SPINK5
2 Diff Krtno: CCL28 (transition cluster towards more differentiated krtno)
3 div : TOP2A
4 sweat gland: VIM, LUM, collagens, APOE
5 sweat gland: OGN, VIM, APOE
6 Corneocyte: KLK1, CNFN, KRT77
7 sweat gland: EBF2, DKK2, LPL, VCAM1
8 bud: PTCH2, LGR6, PTCH1, SEMA3D, LEF1, WNT10B, CTNNB1, KRT14
9 metabolic: ATP6, ND1-6, COX1-3
10 muscle: MYL1/2, TNNT3, TNNC2, TNNI2
11 erythrocyte: ALAS2
12 : PTHLH, SOX9

## Label the clusters
```{r message=FALSE, warning=FALSE}
#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids.sub <- c('0 SOX6 Basal', '1 Diff Krtno', '2 Diff Krtno', '3 Div Krtno', '4 Sweat Gland', '5 Sweat Gland', '6 Corneocyte', '7 Sweat Gland', '8 Basal Bud', '9 Metab Krtno', '10 Muscle', '11 Erythrocyte', '12 PTHLH Sox9 Basal')
names(new.cluster.ids.sub) <- levels(E90_krtno)
E90_krtno <- RenameIdents(E90_krtno, new.cluster.ids.sub)
```

Save the Seurat object
```{r}
save(E90_krtno, file = '~/Desktop/Sean_CellChat/Object/e90_pig_krtno.RData')
```


##Fig Quality DimPlot
```{r fig.height=5, fig.width=7}
DimPlot(E90_krtno, label = TRUE, pt.size = 1.0, label.size = 6,
        cols = c('#0c3695', '#f0b627', '#107de9', '#225ea8', 
                 '#00441b', '#1a9850', '#ff692b', '#66bd63', 
                 '#00206a', '#6b92c3', '#f768a1', '#900101', 
                 '#2c72c8'))#legend
DimPlot(E90_krtno, label = FALSE, pt.size = 1.0,
        cols = c('#0c3695', '#f0b627', '#107de9', '#225ea8', 
                 '#00441b', '#1a9850', '#ff692b', '#66bd63', 
                 '#00206a', '#6b92c3', '#f768a1', '#900101', 
                 '#2c72c8')) + NoLegend()#blank

```

## Recolor the UMAP clusters by cell lineage
```{r fig.height=7, fig.width=7}
path_out <- '~/Desktop/Sean_CellChat/UMAP/'
p1 <- DimPlot(E90_krtno, label = FALSE, pt.size = 1.0, 
              cols = c('#0c3695', '#f0b627', '#107de9', '#225ea8', 
                 '#00441b', '#1a9850', '#ff692b', '#66bd63', 
                 '#00206a', '#6b92c3', '#f768a1', '#900101', 
                 '#2c72c8')) + NoLegend()#manual colors w/o legend
p1
ggsave(plot = p1, filename = 'Pig E90 UMAP (Krtno).png', path = path_out,
         width = 7, height = 7, units = 'in')
```


