---
title: "R Notebook"
output: html_notebook
---

#Load packages
```{r}
library(dplyr)
library(Seurat)
library(tictoc)
library(reticulate)
library(limma)#from Biocmanager
library(ggplot2)
library(viridisLite)
library(patchwork)
library(future)
plan('multisession', workers = 5)#cores = 5
options(future.globals.maxSize = 100000 * 1024^2)

```

#Load data
```{r}
# Load the 6mo dataset
#e30.data <- Read10X_h5("/home/pheonix/Documents/h5files/D30PIG.h5")
#e90.data <- Read10X_h5("/home/pheonix/Documents/h5files/e90pig.h5")
#d10.data <- Read10X_h5("/home/pheonix/Documents/h5files/p10pig.h5")
mon6.data <- Read10X_h5("/home/pheonix/Documents/h5files/6monthpig.h5")

# Initialize the Seurat object with the raw (non-normalized data).
#e30 <- CreateSeuratObject(counts = e30.data, project = "E30", min.cells = 3, min.features = 200)
#e90 <- CreateSeuratObject(counts = e90.data, project = "E90", min.cells = 3, min.features = 200)
#D10 <- CreateSeuratObject(counts = d10.data, project = "P10", min.cells = 3, min.features = 200)
Mon6 <- CreateSeuratObject(counts = mon6.data, project = "6mo", min.cells = 3, min.features = 200)

#e30
#e90
#D10
Mon6

```

#QC
```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
Mon6[["percent.mt"]] <- PercentageFeatureSet(Mon6, pattern = "^MT")

# Visualize QC metrics as a violin plot
VlnPlot(Mon6, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(Mon6, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Mon6, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```

#QC filtering
```{r}
Mon6#before QC filter

VlnPlot(Mon6, features = 'nCount_RNA') + coord_cartesian(ylim = c(0, 10000))
VlnPlot(Mon6, features = 'nFeature_RNA') + coord_cartesian(ylim = c(0, 10000))

Mon6 <- subset(Mon6, subset = 
                 nFeature_RNA > 1000 &
                 nFeature_RNA < 10000 & 
                 percent.mt < 5)

Mon6
```

## SCTransform (Normalize, Scale Data)
```{r, warning=FALSE, message=FALSE}
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
Mon6 <- SCTransform(Mon6, vars.to.regress = "percent.mt", verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
Mon6 <- RunPCA(Mon6, verbose = FALSE)
ElbowPlot(Mon6, ndims = 50)
toc()
```

## UMAP (initial)
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
Mon6 <- RunUMAP(Mon6, dims = 1:40, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
Mon6 <- FindNeighbors(Mon6, dims = 1:40, verbose = FALSE)
Mon6 <- FindClusters(Mon6, verbose = FALSE, algorithm = 3)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(Mon6, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Mon6, label = FALSE, pt.size = 1.0) + NoLegend()
#DimPlot(Mon6, label = FALSE)#no numbers on clusters, legend off to side

toc()
```

## FeaturePlots: view single-cell gene expression across clusters
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
GOI <- c('nCount_RNA',
         'PDGFRA', 'VIM', 'ACTA2', 'LEF1', 'CRABP2',
         'CRABP1', 'RGS5', 'PECAM1', 'LYVE1', 'FLT4', 
         'PTPRC', 'CD86', 'NKG7', 'CDH1',
         'ITGB4', 'ITGA6', 'SOX6', 'KRT15', 'KRT14', 'KRT5', 'KRT1', 'KRT10', 'KRT8', 'CDSN', 
         'TOP2A', 'SOX10', 'DCT', 'SOX9', 'DLK2', 'DLL1', 'PDGFA', 'VEGFA', 'ANGPTL1', 'ANGPTL2')
#path_out <- '.'
## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = Mon6,
                    features = GOI[n_plot],
                    pt.size = 1,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
  }
toc()
```

### Change Resolution
```{r message=FALSE, warning=FALSE}
Mon6 <- FindClusters(Mon6, verbose = FALSE, algorithm = 3, resolution = 0.7)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(Mon6, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Mon6, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

## Differential Gene Expression
```{r}
tic('ClusterMarker auto-export loop')#start timer
path_out <- '~/Desktop/Sean_CellChat/DEG/6mo/'#path to export folder
COI <- levels(Mon6)#define the clusters you wish to know diff. gene expression of

n_genes <- length(Mon6@assays[["SCT"]]@data@Dimnames[[1]])#get # of genes we will be testing
n_loops <- 1
while (n_loops < length(COI) + 1) {
#Find positive markers (upregulated genes)
cluster.markers.temp <- FindMarkers(Mon6, ident.1 = COI[n_loops], min.pct = 0.05, logfc.threshold = 0.0, assay = 'SCT', only.pos = TRUE)#iterate through the entire COI list
#head(cluster.markers.temp, n = 30)
cluster.markers.temp$Gene = rownames(cluster.markers.temp)#add a column for the gene name
cluster.markers.temp$pct.2[cluster.markers.temp$pct.2 == 0] <- NA#corrects pct ratio error if pct.2=0
cluster.markers.temp <- na.omit(cluster.markers.temp)#remove rows with NA
cluster.markers.temp$Pct_Ratio = cluster.markers.temp$pct.1 / cluster.markers.temp$pct.2#compute ratio of how much the cluster of interest expresses compared to all other cells
cluster.markers.temp <- cluster.markers.temp %>% dplyr::select('Gene', everything())#move gene column to the front of the dataframe
cluster.markers.temp$'PctRatio_x_logfc' = cluster.markers.temp$Pct_Ratio * cluster.markers.temp$avg_log2FC
cluster.markers.temp <- cluster.markers.temp[cluster.markers.temp$PctRatio_x_logfc > quantile(cluster.markers.temp$PctRatio_x_logfc, 0.95), ]#keep only the top 95% of upregulated genes in ident.1
cluster.markers.temp <- cluster.markers.temp[order(cluster.markers.temp$PctRatio_x_logfc, decreasing = TRUE), ]#rank genes
write.csv(cluster.markers.temp, file = paste(path_out, (n_loops-1), '_pos_markers.csv', sep = ''), row.names = FALSE)#save that cluster's markers

n_loops <- n_loops + 1#track for when all clusters of one dataset have been run through
}
toc()#end timer
```

#View Top DEGs
```{r fig.height=20, fig.width=14}
path_in <- '~/Desktop/Sean_CellChat/DEG/6mo/'#path to import folder

DimPlot(Mon6, label = TRUE, pt.size = 3, label.size = 9.0)

#load DEGs
temp_clusters <- levels(Mon6$seurat_clusters)
n_cluster <- 0
while (n_cluster < length(temp_clusters)) {
  temp.csv <- read.csv(file = paste(path_in, n_cluster, '_pos_markers.csv', sep = ''))#read in csv file of DEGs
  temp.csv <- temp.csv[temp.csv$pct.1 >= 0.5, ]#filter to just DEGs expressed in greater than 50% of the cells in the cluster
  temp.csv <- temp.csv[1:10, ]#keep top 10 DEGs
  temp_degs <- temp.csv$Gene#convert top 10 DEGs to a string vector
  
  if (n_cluster == 0) {
    degs <- temp_degs#start storage variable w/ first cluster's top 10 degs
  }
  else {
    degs <- c(degs, temp_degs)#append the cluster's top 10 degs to the storage variable
  }
  
  n_cluster <- n_cluster + 1#iterate to next cluster
}

Mon6degs <- DoHeatmap(Mon6,
                features = degs,
                label = TRUE,
                slot = 'data',
                disp.max = 4) +
  scale_fill_gradientn(colours = magma(50))
print(Mon6degs)

Mon6degs <- DoHeatmap(Mon6,
                features = c('PDGFRA', 'ITGA6', 'KRT10', 'EDAR', 'LEF1', 'RGS5', 'PTPRC', 'ACTA2', 'PPARG'),
                label = TRUE,
                slot = 'data',
                disp.max = 4) +
  scale_fill_gradientn(colours = magma(50))
print(Mon6degs)
```


## Label the Clusters by Cell Type
```{r message=FALSE, warning=FALSE}
#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids <- c('0 Pericyte', '1 Basal Krtno', '2 Krtno', '3 Fibroblast', '4 Krtno',
                     '5 K1_K10 Krtno', '6 BV', '7 Krtno', '8 Div Krtno', '9 Immune Cell',
                     '10 Macrophage', '11 CDH1 Krtno', '12 Basal Krtno', '13 Immune Cell', '14 Macrophage',
                     '15 Schwann', '16 Lymphatic', '17 Corneocyte')
names(new.cluster.ids) <- levels(Mon6)
Mon6 <- RenameIdents(Mon6, new.cluster.ids)
```

Save the Seurat object
```{r}
save(Mon6, file = '~/Desktop/Sean_CellChat/Object/6mo_pig_allclusters.RData')
```

## Recolor the UMAP clusters by cell lineage
```{r fig.height=7, fig.width=7}
path_out <- '~/Desktop/Sean_CellChat/UMAP/'
p1 <- DimPlot(Mon6, label = FALSE, pt.size = 0.5, 
              cols = c('#7fcdbb', '#002783', '#64a7e9', '#238b45', '#4d8ad5',
                     '#96ccff', '#8c6bb1', '#0c3695', '#2565b5', '#BB0000',
                     '#790001', '#16539e', '#00206a', '#dc0000', '#900101',
                     '#ffab4b', '#88419d', '#9decff')) + NoLegend()#manual colors w/o legend
p1
ggsave(plot = p1, filename = 'Pig 6mo UMAP (All Clusters).png', path = path_out,
         width = 7, height = 7, units = 'in')
```


# 3. Subset the Keratinocytes
```{r message=FALSE, warning=FALSE}
Mon6_krtno <- subset(Mon6, idents = c('1 Basal Krtno', '2 Krtno', '4 Krtno', '5 K1_K10 Krtno', '7 Krtno', '8 Div Krtno', 
                     '11 CDH1 Krtno', '12 Basal Krtno', '17 Corneocyte'), invert = FALSE)#subset the keratinocyte supercluster
Mon6_krtno#view the subset

## We want to renormalize the data to bring out the heterogenetiy within the lineage
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
Mon6_krtno <- SCTransform(Mon6_krtno, vars.to.regress = "percent.mt", verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
Mon6_krtno <- RunPCA(Mon6_krtno, verbose = FALSE)
ElbowPlot(Mon6_krtno, ndims = 50)
```

Based on ElbowPlot, pick major PCs for next steps
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
Mon6_krtno <- RunUMAP(Mon6_krtno, dims = 1:30, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
Mon6_krtno <- FindNeighbors(Mon6_krtno, dims = 1:30, verbose = FALSE)
Mon6_krtno <- FindClusters(Mon6_krtno, verbose = FALSE, algorithm = 3, resolution = 0.6)#default is algorithm = 1 (Louvain), 3 = SLM
toc()
DimPlot(Mon6_krtno, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Mon6_krtno, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

### View Gene Expression of Keratinocyte Heterogeneity Markers
Includes some markers identified by Sole-Boldo et al. in their analysis as well as DAGs from Thompson et al. 2022
```{r fig.height=5, fig.width=7}
## FeaturePlots: view single-cell gene expression across clusters
DimPlot(Mon6_krtno, label = TRUE, pt.size = 2, label.size = 6.0)

GOI <- c('PDGFRA', 'APOE', 'KRT14', 'KRT5', 'KRT10', 'KRT18', 'KRT8', 'CALML5', 'CDSN', 'CNFN', 'CDH1',
         'TOP2A', 'ASPM', 'ITGA6', 'ITGB1', 'SOX6', 'SOSTDC1', 'KRT15', 'COL17A1', 'SOX9', 'VEGFA', 'KRT10', 'nFeature_SCT', 'TGFBI', 'DPP4', 'ACTA2', 'PPARG', 'LEF1',
         'DLL1', 'CAV1', 'DUSP6', 'COL17A1', 'CTNNB1')

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = Mon6_krtno,
                    features = GOI[n_plot],
                    pt.size = 1.0,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()

```

### Change Resolution
```{r message=FALSE, warning=FALSE}
Mon6_krtno <- FindClusters(Mon6_krtno, verbose = FALSE, algorithm = 3, resolution = 0.6)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(Mon6_krtno, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Mon6_krtno, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

# Differential Gene Expression Within Keratinocytes
```{r}
tic('ClusterMarker auto-export loop')#start timer
path_out <- '~/Desktop/Sean_CellChat/DEG/Krtno_sub/6mo/'#path to export folder
COI <- c('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11')#define the clusters you wish to know diff. gene expression of

n_genes <- length(Mon6_krtno@assays[["SCT"]]@data@Dimnames[[1]])#get # of genes we will be testing
n_loops <- 1
while (n_loops < length(COI) + 1) {
#Find positive markers (upregulated genes)
cluster.markers.temp <- FindMarkers(Mon6_krtno, ident.1 = COI[n_loops], min.pct = 0.05, logfc.threshold = 0.0, assay = 'SCT', only.pos = TRUE)#iterate through the entire COI list
#head(cluster.markers.temp, n = 30)
cluster.markers.temp$Gene = rownames(cluster.markers.temp)#add a column for the gene name
cluster.markers.temp$pct.2[cluster.markers.temp$pct.2 == 0] <- NA#corrects pct ratio error if pct.2=0
cluster.markers.temp <- na.omit(cluster.markers.temp)#remove rows with NA
cluster.markers.temp$Pct_Ratio = cluster.markers.temp$pct.1 / cluster.markers.temp$pct.2#compute ratio of how much the cluster of interest expresses compared to all other cells
cluster.markers.temp <- cluster.markers.temp %>% dplyr::select('Gene', everything())#move gene column to the front of the dataframe
cluster.markers.temp$'PctRatio_x_logfc' = cluster.markers.temp$Pct_Ratio * cluster.markers.temp$avg_log2FC
cluster.markers.temp <- cluster.markers.temp[cluster.markers.temp$PctRatio_x_logfc > quantile(cluster.markers.temp$PctRatio_x_logfc, 0.95), ]#keep only the top 95% of upregulated genes in ident.1
cluster.markers.temp <- cluster.markers.temp[order(cluster.markers.temp$PctRatio_x_logfc, decreasing = TRUE), ]#rank genes
write.csv(cluster.markers.temp, file = paste(path_out, COI[n_loops], '_pos_markers.csv', sep = ''), row.names = FALSE)#save that cluster's markers

n_loops <- n_loops + 1#track for when all clusters of one dataset have been run through
}
toc()#end timer
```
0 Diff: KRT1/10, LRATD1
1 Diff: TPRG1, MBOAT2, ENOX1, TEC, SLC9A9, LRP4, GAS7, K10/CALML5 low
2 Basal: KRT14/5, CDH1, ; , CCL2/20, IFIH1
3 Diff: KRT2, SERPINB12, CALML5, CNFN, KRT10
4 Rete: PRSS23, ITGB4, KRT5, PCDH7, KRT14, COL17A1, SOX6, COL4A5/6
5 Rete: DLK2, COCH, KRT15, IL1R2, SPON1, PDPN, SMOC2, COL12A1, ITGA2, COL14A1, COL5A2
6 Div: TOP2A, CENPA/F
7 : HS3ST1, MMP25, AREG, SOWAHB, EREG, PRDM1, PLAUR, FLRT3, ADAMTS17
8 Sweat Gland: EBF1, FN1, MYO1B, COL4A2, COL6A1, VIM
9 : ZFPM2, DACH1, SOX9, ADGRV1, PRICKLE2, HOMER1, RUNX1, PARP8, MYLK
10 : KLK1, KRT77, ALOX12B, CDSN, CD24, CNFN
11 : TENM2, PCDH7, SOX6, ITGB4, ROBO2, ITGA2, COL4A6, PDGFC, COL4A5

## Label the clusters
```{r message=FALSE, warning=FALSE}
#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids.sub <- c('0 Diff Krtno', '1 Diff Krtno', '2 Basal Krtno', '3 Diff Krtno', 
                         '4 Rete Krtno', '5 Rete Krtno', '6 Div Krtno', '7 AREG Krtno', 
                         '8 Sweat Gland', '9 SOX9 Krtno', '10 Corneocyte', '11 Rete Krtno')
names(new.cluster.ids.sub) <- levels(Mon6_krtno)
Mon6_krtno <- RenameIdents(Mon6_krtno, new.cluster.ids.sub)
```

Save the Seurat object
```{r}
save(Mon6_krtno, file = '~/Desktop/Sean_CellChat/Object/6mo_pig_krtno.RData')
```


##Fig Quality DimPlot
```{r fig.height=5, fig.width=7}
DimPlot(Mon6_krtno, label = TRUE, pt.size = 1.0, label.size = 6,
        cols = c('#f09527', '#f0b627', '#107de9', '#ff692b', 
                 '#2c72c8', '#0c3695', '#225ea8', '#8c6bb1', 
                 '#1a9850', '#00206a', '#d04800', '#6b92c3'))#legend
DimPlot(Mon6_krtno, label = FALSE, pt.size = 1.0,
        cols = c('#f09527', '#f0b627', '#107de9', '#ff692b', 
                 '#2c72c8', '#0c3695', '#225ea8', '#8c6bb1', 
                 '#1a9850', '#00206a', '#d04800', '#6b92c3')) + NoLegend()#blank

```

## Recolor the UMAP clusters by cell lineage
```{r fig.height=7, fig.width=7}
path_out <- '~/Desktop/Sean_CellChat/UMAP/'
p1 <- DimPlot(Mon6_krtno, label = FALSE, pt.size = 1.0, 
              cols = c('#f09527', '#f0b627', '#107de9', '#ff692b', 
                 '#2c72c8', '#0c3695', '#225ea8', '#8c6bb1', 
                 '#1a9850', '#00206a', '#d04800', '#6b92c3')) + NoLegend()#manual colors w/o legend
p1
ggsave(plot = p1, filename = 'Pig 6mo UMAP (Krtno).png', path = path_out,
         width = 7, height = 7, units = 'in')
```

