---
title: "Sean_Glover2023_Reanalysis"
output: html_notebook
---

#0. Load packages
```{r}
library(dplyr)
library(Seurat)
library(tictoc)
library(reticulate)
library(limma)#from Biocmanager
library(ggplot2)
library(viridisLite)
library(patchwork)
library(future)
library(tictoc)
plan('multisession', workers = 5)#cores = 5
options(future.globals.maxSize = 100000 * 1024^2)

#install.packages('textshaping')#needed ~"sudo apt install libharfbuzz-dev libfribidi-dev" first
#install.packages('Cairo)#needed ~"sudo apt install libxt-dev", ~"sudo apt-get install libcairo2-dev" first
#install.packages('units')#needed ~"sudo apt-get install libudunits2-dev" first
#install.packages('ragg')#needed ~"libfreetype6-dev libpng-dev libtiff5-dev libjpeg-dev" first
#install.packages('terra')#needed ~"sudo apt-get install gdal-bin libgdal-dev" first
#install.packages('sf')#
#install.packages('spdep')#
#install.packages('ggrastr')#

library(monocle3)#installed via devtools::install_github('cole-trapnell-lab/monocle3'); Monocle3 dependencies listed above
library(Seurat)
library(SeuratWrappers)#installed via remotes::install_github('satijalab/seurat-wrappers')
library(patchwork)
library(ggplot2)
#library(dplyr)
set.seed(1234)
```

#Load data
```{r}
# Load the D10 dataset
donor1.data <- Read10X("~/Documents/Glover2023/VentralHand1(GSM5844682)/")
donor2.data <- Read10X("~/Documents/Glover2023/VentralHand2(GSM5844685)/")

# Initialize the Seurat object with the raw (non-normalized data).
donor1 <- CreateSeuratObject(counts = donor1.data, project = "Ventral1", min.cells = 3, min.features = 200)
donor2 <- CreateSeuratObject(counts = donor2.data, project = "Ventral2", min.cells = 3, min.features = 200)
  
donor1
donor2

Glover2023 <- merge(x = donor1, y = c(donor2), add.cell.ids = c('V1', 'V2'), project = 'Glover2023_Human_Ventral')
```

#QC
```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
Glover2023[["percent.mt"]] <- PercentageFeatureSet(Glover2023, pattern = "^MT")

# Visualize QC metrics as a violin plot
VlnPlot(Glover2023, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(Glover2023, features = c("nCount_RNA")) + ylim(0, 10000)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(Glover2023, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Glover2023, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```

#QC Filter
```{r}
Glover2023 <- subset(x = Glover2023, 
                          subset = nCount_RNA > 1250 &
                            nCount_RNA < 10000)
Glover2023

# Visualize QC metrics as a violin plot
VlnPlot(Glover2023, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

## SCTransform (Normalize, Scale Data)
```{r, warning=FALSE, message=FALSE}
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
Glover2023 <- SCTransform(Glover2023, vars.to.regress = "percent.mt", verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
Glover2023 <- RunPCA(Glover2023, verbose = FALSE)
ElbowPlot(Glover2023, ndims = 50)
```

## UMAP (initial)
Pick dimensions to use based on ElbowPlot above -- try to pick dimensions w/ > 1 standard deviation. In this case, up to 40 dimensions is greater than 5 standard deviations before it levels off (suggesting decline in variability).
```{r message=FALSE, warning=FALSE}
tic('Running UMAP & Making DimPlots')
#run the UMAP function using dimensions informed by elbow plot
Glover2023 <- RunUMAP(Glover2023, dims = 1:40, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
Glover2023 <- FindNeighbors(Glover2023, dims = 1:40, verbose = FALSE)
Glover2023 <- FindClusters(Glover2023, verbose = FALSE, algorithm = 3)#default is algorithm = 1/2 (Louvain), 3 = SLM (scATAC), 4 = Leiden (not installed on server currently - need to pip install leidenalg later)
DimPlot(Glover2023, label = TRUE, pt.size = 0.5, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Glover2023, label = FALSE, pt.size = 0.5) + NoLegend()#no labels (to label in photoshop)
DimPlot(Glover2023, group.by = 'orig.ident', pt.size = 0.5)
#DimPlot(Glover2023, label = FALSE)#no numbers on clusters, legend off to side
toc()

#Check QC Metrics on Clustering
FeaturePlot(Glover2023, features = c('nFeature_RNA', 'nCount_RNA'))
FeaturePlot(Glover2023, features = c('percent.mt'))
```

#FeaturePlot
```{r message=FALSE, warning=FALSE}
## FeaturePlots: view single-cell gene expression across clusters
GOI <- c("PDGFRA", "FGF20", "PECAM1", "LYVE1", "SOX10", "PTPRC", "CD86", "KRT14", "KRT10", "KRT8", "KRT18", "KRT15", 'CALML5', 'CNFN', "ITGA6", "ITGB4", 
         "LEF1", "EDAR", "WNT10B", "PTCH1", "PTCH2", "VIM", "SOX6", "SOX9", "COL17A1", "PTPRC", "ALAS2", "VEGFA", "COL1A1", "ACTA2", "TOP2A", "PDGFA", "EDA", "TGFA", "LMX1A", "RGS5")

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = Glover2023,
                    features = GOI[n_plot],
                    pt.size = 0.3,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()
```

### Change Resolution
```{r fig.height=5, fig.width=5, message=FALSE, warning=FALSE}
Glover2023 <- FindClusters(Glover2023, verbose = FALSE, algorithm = 3, resolution = 0.2)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(Glover2023, label = TRUE, pt.size = 0.5, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Glover2023, label = FALSE, pt.size = 0.5) + NoLegend()#no labels (to label in photoshop)
DimPlot(Glover2023, label = FALSE, pt.size = 0.5, group.by = 'orig.ident')#no labels (to label in photoshop)
```

Save the R object
```{r}
save(Glover2023, file = '~/Desktop/Sean_CellChat/Object/Glover2023_Ventral_AllClusters.RData')#final object w/ labels
```

## Differential Gene Expression
```{r}
tic('ClusterMarker auto-export loop')#start timer
path_out <- '~/Desktop/Sean_CellChat/DEG/Glover2023/'#path to export folder
COI <- levels(Glover2023)#define the clusters you wish to know diff. gene expression of

n_genes <- length(Glover2023@assays[["SCT"]]@data@Dimnames[[1]])#get # of genes we will be testing
n_loops <- 1
while (n_loops < length(COI) + 1) {
#Find positive markers (upregulated genes)
cluster.markers.temp <- FindMarkers(Glover2023, ident.1 = COI[n_loops], min.pct = 0.05, logfc.threshold = 0.0, assay = 'SCT', only.pos = TRUE)#iterate through the entire COI list
#head(cluster.markers.temp, n = 30)
cluster.markers.temp$Gene = rownames(cluster.markers.temp)#add a column for the gene name
cluster.markers.temp$pct.2[cluster.markers.temp$pct.2 == 0] <- NA#corrects pct ratio error if pct.2=0
cluster.markers.temp <- na.omit(cluster.markers.temp)#remove rows with NA
cluster.markers.temp$Pct_Ratio = cluster.markers.temp$pct.1 / cluster.markers.temp$pct.2#compute ratio of how much the cluster of interest expresses compared to all other cells
cluster.markers.temp <- cluster.markers.temp %>% dplyr::select('Gene', everything())#move gene column to the front of the dataframe
cluster.markers.temp$'PctRatio_x_logfc' = cluster.markers.temp$Pct_Ratio * cluster.markers.temp$avg_log2FC
cluster.markers.temp <- cluster.markers.temp[cluster.markers.temp$PctRatio_x_logfc > quantile(cluster.markers.temp$PctRatio_x_logfc, 0.95), ]#keep only the top 95% of upregulated genes in ident.1
cluster.markers.temp <- cluster.markers.temp[order(cluster.markers.temp$PctRatio_x_logfc, decreasing = TRUE), ]#rank genes
write.csv(cluster.markers.temp, file = paste(path_out, (n_loops-1), '_pos_markers.csv', sep = ''), row.names = FALSE)#save that cluster's markers

n_loops <- n_loops + 1#track for when all clusters of one dataset have been run through
}
toc()#end timer
```
10: KRT20; KRT8/18 also expressed (https://pubmed.ncbi.nlm.nih.gov/7769257/ - Merkel Cells)


#View Top DEGs
```{r fig.height=20, fig.width=14}
path_in <- '~/Desktop/Sean_CellChat/DEG/Glover2023/'#path to import folder

DimPlot(Glover2023, label = TRUE, pt.size = 3, label.size = 9.0)

#load DEGs
temp_clusters <- levels(Glover2023$seurat_clusters)
n_cluster <- 0
while (n_cluster < length(temp_clusters)) {
  temp.csv <- read.csv(file = paste(path_in, n_cluster, '_pos_markers.csv', sep = ''))#read in csv file of DEGs
  temp.csv <- temp.csv[temp.csv$pct.1 >= 0.5, ]#filter to just DEGs expressed in greater than 50% of the cells in the cluster
  temp.csv <- temp.csv[1:10, ]#keep top 10 DEGs
  temp_degs <- temp.csv$Gene#convert top 10 DEGs to a string vector
  
  if (n_cluster == 0) {
    degs <- temp_degs#start storage variable w/ first cluster's top 10 degs
  }
  else {
    degs <- c(degs, temp_degs)#append the cluster's top 10 degs to the storage variable
  }
  
  n_cluster <- n_cluster + 1#iterate to next cluster
}

p1 <- DoHeatmap(Glover2023,
                features = degs,
                label = TRUE,
                slot = 'data',
                disp.max = 4) +
  scale_fill_gradientn(colours = magma(50))
print(p1)

p1 <- DoHeatmap(Glover2023,
                features = c('PDGFRA', 'ITGA6', 'KRT14', 'KRT5', 'KRT1', 'KRT10', 'EDAR', 'LEF1', 'RGS5', 'PTPRC', 'ACTA2', 'PPARG'),
                label = TRUE,
                slot = 'data',
                disp.max = 4) +
  scale_fill_gradientn(colours = magma(50))
print(p1)
```


## Recolor the UMAP clusters by cell lineage
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
path_out <- '~/Desktop/Sean_CellChat/UMAP/'
p1 <- DimPlot(Glover2023, label = FALSE, pt.size = 1.0, 
              cols = c('#00441b', '#0c3695', '#225ea8', '#41ab5d', '#7fcdbb', 
                       '#ffab4b', '#006d2c', '#238b45', '#8c6bb1', '#640001', 
                       '#fc8d59', '#BB0000', '#88419d')) + NoLegend()#manual colors w/o legend
p1
ggsave(plot = p1, filename = 'Glover2023 Human GW UMAP (All Clusters).png', path = path_out,
         width = 7, height = 7, units = 'in')

```


#2. Subset Keratinocytes
```{r}
Glover2023_krtno <- subset(Glover2023, idents = c('1', '2'), invert = FALSE)#subset the keratinocyte clusters
Glover2023_krtno#view the subset
```

#scTransform
```{r}
tic('Running SCTransform')
Glover2023_krtno <- SCTransform(Glover2023_krtno, vars.to.regress = "percent.mt", verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
Glover2023_krtno <- RunPCA(Glover2023_krtno, verbose = FALSE)
ElbowPlot(Glover2023_krtno, ndims = 50)
```

Based on ElbowPlot, pick major PCs for next steps
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
Glover2023_krtno <- RunUMAP(Glover2023_krtno, dims = 1:30, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
Glover2023_krtno <- FindNeighbors(Glover2023_krtno, dims = 1:30, verbose = FALSE)
Glover2023_krtno <- FindClusters(Glover2023_krtno, verbose = FALSE, algorithm = 3, resolution = 0.6)#default is algorithm = 1 (Louvain), 3 = SLM
toc()
DimPlot(Glover2023_krtno, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Glover2023_krtno, label = FALSE, pt.size = 1.0) + NoLegend()#no labels
DimPlot(Glover2023_krtno, label = FALSE, pt.size = 1.0, group.by = 'orig.ident')#colored by sample
```

#FeaturePlot
```{r message=FALSE, warning=FALSE}
## FeaturePlots: view single-cell gene expression across clusters
GOI <- c("FGF20", "KRT14", "KRT10", "CNFN", "KRT15", "ITGA6", "ITGB4", 
         "LEF1", "EDAR", "WNT10B", "PTCH1", "PTCH2", "SOX6", "SOX9", "COL17A1", "VEGFA", "TOP2A", "PDGFA", "EDA", "EN1", "TGFA", "LMX1A", "KRT8", "KRT18", "KRT19")

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = Glover2023_krtno,
                    features = GOI[n_plot],
                    pt.size = 0.5,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()
```

### Change Resolution
```{r message=FALSE, warning=FALSE}
Glover2023_krtno <- FindClusters(Glover2023_krtno, verbose = FALSE, algorithm = 3, resolution = 0.3)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(Glover2023_krtno, label = TRUE, pt.size = 0.5, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Glover2023_krtno, label = FALSE, pt.size = 0.5) + NoLegend()#no labels (to label in photoshop)
DimPlot(Glover2023_krtno, label = FALSE, pt.size = 0.5, group.by = 'orig.ident')#no labels (to label in photoshop)
```

##Figure quality UMAP
```{r fig.height=7, fig.width=7}
path_out <- '~/Desktop/Sean_CellChat/UMAP/'
DimPlot(Glover2023_krtno, label = TRUE, pt.size = 2, label.size = 6,
        cols = c('#0c3695', '#00206a', '#f0b627', '#ff692b', '#4d8ad5', '#8c6bb1'))#legend
p1 <- DimPlot(Glover2023_krtno, label = FALSE, pt.size = 2,
        cols = c('#0c3695', '#00206a', '#f0b627', '#ff692b', '#4d8ad5', '#8c6bb1')) + NoLegend()#blank
p1

ggsave(plot = p1, filename = 'Glover2023 Human Ventral1_2 UMAP (Keratinocytes).png', path = path_out,
         width = 7, height = 7, units = 'in')

```

Save the R object
```{r}
save(Glover2023_krtno, file = '/t2tb/Object/Glover2023_Ventral_Krtno.RData')
```

#Convert Seurat to CDS and Do Pseudotime in Monocle3
```{r}
#want to transfer Seurat UMAP coordinates to CDS object so pseudotime is on the Seurat clusters
temp1 <- Glover2023_krtno@reductions$umap@cell.embeddings#large matrix with rows cell barcodes and 2 columns for the Seurat x/y UMAP coordinates
temp2 <- Glover2023_krtno@meta.data$SCT_snn_res.0.3#factor levels for the cluster ids at the resolution used to cluster the Seurat object

cds_glover <- as.cell_data_set(Glover2023_krtno, group.by = 'SCT_snn_res.0.3')#function from SeuratWrappers to convert to Monocle3 object, graph argument specifies the clustering to transfer (assay to use, and active resolution) to @clusters
#cds_glover <- cluster_cells(cds_glover, reduction_method = "UMAP", group_cells_by = 'seurat_clusters')#

#Add some info not carried over by SeuratWrappers
cds_glover <- estimate_size_factors(cds_glover)#add size factor metadata to CDS object
cds_glover@rowRanges@elementMetadata@listData[["gene_short_name"]] <- rownames(Glover2023_krtno)#correct gene names in CDS object

#Validate UMAP coordinate transfer from Seurat object -> CDS object
temp1 <- Glover2023_krtno@reductions$umap@cell.embeddings#large matrix with rows cell barcodes and 2 columns for the Seurat x/y UMAP coordinates
temp2 <- cds_glover@int_colData@listData$reducedDims$UMAP#matrix with rows cell barcodes and 2 columns for x/y UMAP coordinates
table(temp1 == temp2)#if all TRUE, seuratwrapper successfully transferred umap coordinates to CDS object via the graph argument
#cds_glover@int_colData@listData$reducedDims$UMAP <- Glover2023_krtno@reductions$umap@cell.embeddings#not needed due to graph argument when making CDS object

#need to transfer Seurat PCA since this wasn't transferred by seurat wrapper
reducedDim(cds_glover, type = "PCA") <- Glover2023_krtno@reductions$pca@cell.embeddings#transfer the matrix results of Seurat PCA to CDS
#cds_glover@preprocess_aux$prop_var_expl <- Glover2023_krtno@reductions$pca@stdev#transfer numerical vector containing stdev values per PCA dimension

#partitions still need to be created for learn_graph()
#manual_partition <- Glover2023_krtno@meta.data$seurat_clusters#get numerical vector of cluster id #s to construct manual partition
#krt_id <- c('0', '1', '2', '3', '6', '8', '9', '12')#keratinocytes
#misc_id <- c('4', '5', '7', '10', '11')#non-keratinocytes
#krt_i <- grep(pattern = krt_id, x = manual_partition)#find index position of matching cluster ids
#misc_i <- grep(pattern = misc_id, x = manual_partition)#find index position of matching cluster ids
#manual_partion[krt_i] <- '1'#set all krtno cluster ids to manual partition id '1'
#manual_partition[misc_i] <- '2'#set non-krtno cluster ids to manual partition id '2'
#table(levels(manual_partition))#validate replacement

manual_partition <- as.vector(Glover2023_krtno@meta.data$seurat_clusters)#get cluster id #s to construct manual partition from a factor
names(manual_partition) <- rownames(Glover2023_krtno@meta.data)#
manual_partition <- as.factor(manual_partition)
head(manual_partition)#view ordered levels
levels(manual_partition) <- c(1, 1, 1, 1, 1, 1)#replace cluster id levels with new ids for partitioning in Monocle3
table(levels(manual_partition))#validate replacement
cds_glover@clusters@listData$UMAP$partitions <- manual_partition


#recreate.partition <- c(rep(1, length(cds_glover@colData@rownames)))
#names(recreate.partition) <- cds_glover@colData@rownames#assigns partition value to cell barcode
#recreate.partition <- as.factor(recreate.partition)#converts this to a large factor with 1 level
#cds_glover@clusters@listData$UMAP$partitions <- recreate.partition

p1 <- plot_cells(cds_glover, reduction_method = 'UMAP',
                 color_cells_by = 'seurat_clusters', 
                 label_cell_groups = FALSE,
                 cell_size = 1,
                 show_trajectory_graph = FALSE)#is a ggplot2 object, so can custom color same as with Seurat's DimPlot
p1
plot_cells(cds_glover, reduction_method = 'UMAP',
                 color_cells_by = 'cluster', 
                 label_cell_groups = FALSE,
                 cell_size = 1,
                 show_trajectory_graph = FALSE)#Monocle3 clusters, transferred from Seurat
plot_cells(cds_glover, reduction_method = 'UMAP',
                 color_cells_by = 'partition', 
                 label_cell_groups = TRUE,
                 cell_size = 1,
                 show_trajectory_graph = FALSE)
#pseudotime

#rownames(cds_glover@principal_graph_aux$UMAP$dp_mst) <- NULL
#colnames(cds_glover@int_colData@listData$reducedDims$UMAP) <- NULL

cds_glover <- learn_graph(cds_glover, use_partition = TRUE, verbose = TRUE)
p2 <- plot_cells(cds_glover, reduction_method = 'UMAP',
                 color_cells_by = 'seurat_clusters', 
                 group_label_size = 5, cell_size = 1,
                 show_trajectory_graph = TRUE)#is a ggplot2 object, so can custom color same as with Seurat's DimPlot
p2

plot_cells(cds_glover, reduction_method = 'UMAP',
                 color_cells_by = 'partition', 
                 group_label_size = 5, cell_size = 1,
                 show_trajectory_graph = TRUE)#is a ggplot2 object, so can custom color same as with Seurat's DimPlot


cds_glover <- order_cells(cds_glover)#choose root node
plot_cells(cds_glover,
           color_cells_by = "pseudotime",
           group_cells_by = "cluster",
           label_cell_groups = FALSE,
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           label_roots = FALSE,
           cell_size = 1,
           trajectory_graph_color = "grey60")
plot_cells(cds_glover,
           color_cells_by = "pseudotime",
           group_cells_by = "cluster",
           label_cell_groups = FALSE,
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           label_roots = FALSE,
           cell_size = 1,
           show_trajectory_graph = FALSE)
```

Save CDS object
```{r}
save(cds_glover, file = '/t2tb/Object/Glover2023_ventral_krtno_CDS.RData')
```

#View Expression Trajectories Across Keratinocyte Clusters (w/ GAM)
```{r}
pseudotime_df <- as.data.frame(t(cds_glover@assays@data@listData$logcounts))#get cell by gene expression matrix
pseudotime_df$Barcode <- rownames(pseudotime_df)#get column of barcode values
pseudotime_df$Pseudotime <- pseudotime(cds_glover)#get pseudotime values

temp_df <- data.frame("Barcode" = rownames(as.data.frame(cds_glover@clusters@listData$UMAP$clusters)), 
                      "Cluster" = cds_glover@clusters@listData$UMAP$clusters)#get cluster ids

pseudotime_df <- merge(temp_df, pseudotime_df)#filter to shared rows in subset cds
pseudotime_df <- pseudotime_df[pseudotime_df$Pseudotime != Inf, ]#remove rows w/ inf pseudotime value
#write.csv(pseudotime_df, file = '~/Desktop/Sean_CellChat/Pseudotime/Glover2023_Ventral_Krtno_Pseudotime.csv', row.names = FALSE)#save pseudotime data

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey40', se = FALSE) +
  stat_smooth(aes(y = PTCH1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue', se = FALSE) +
  stat_smooth(aes(y = PTCH2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +  
  #stat_smooth(aes(y = SHH), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +   
  stat_smooth(aes(y = WNT10B), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red3', se = FALSE) +
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red4', se = FALSE) +
  stat_smooth(aes(y = EDA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid1', se = FALSE) +   
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal Bud DEGs")#SHH not detected in this dataset

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey40', se = FALSE) +
  #stat_smooth(aes(y = SHH), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +   
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red4', se = FALSE) +
  #stat_smooth(aes(y = EDA), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'orchid1', se = FALSE) +   
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal Bud")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey60', se = FALSE, linetype = 'dashed') +
  #stat_smooth(aes(y = SHH), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'blue', se = FALSE) +   
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red', se = FALSE) +
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'green', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal Bud (Figure)")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = KRT15), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +
  stat_smooth(aes(y = KRT14), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue3', se = FALSE) +  
  stat_smooth(aes(y = KRT5), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +
  stat_smooth(aes(y = KRT10), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod2', se = FALSE) +
  stat_smooth(aes(y = KRT1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkorange', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Keratins")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey30', se = FALSE) +
  stat_smooth(aes(y = ITGA2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey50', se = FALSE) +
  stat_smooth(aes(y = SOX6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +
  stat_smooth(aes(y = SOX9), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'peru', se = FALSE) +  
  stat_smooth(aes(y = DLK2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue3', se = FALSE) +
  stat_smooth(aes(y = NRG1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +
  stat_smooth(aes(y = PHLDB2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  stat_smooth(aes(y = LAMB3), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid4', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal IFE DEGs")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = VEGFA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid1', se = FALSE) +
    stat_smooth(aes(y = VEGFC), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  stat_smooth(aes(y = ANGPTL1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple1', se = FALSE) +
  stat_smooth(aes(y = ANGPTL2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple3', se = FALSE) +
  stat_smooth(aes(y = ANGPTL4), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple4', se = FALSE) +
  stat_smooth(aes(y = PDGFA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgreen', se = FALSE) +
  stat_smooth(aes(y = PDGFB), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'lightseagreen', se = FALSE) +  
  stat_smooth(aes(y = PDGFC), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'green3', se = FALSE) +
  #stat_smooth(aes(y = COCH), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'darkgreen', se = FALSE) +
  #stat_smooth(aes(y = COL2A1), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'lightseagreen', se = FALSE) +  
  theme_classic() +
  ggtitle("Pseudotime Dermal Pocket Genes")


ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = RDH10), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid1', se = FALSE) +
  stat_smooth(aes(y = DHRS3), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkred', se = FALSE) +
  stat_smooth(aes(y = ALDH1A1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple3', se = FALSE) +
  stat_smooth(aes(y = ALDH1A3), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple4', se = FALSE) +
  stat_smooth(aes(y = RARA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +
  stat_smooth(aes(y = RARG), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue3', se = FALSE) +
  stat_smooth(aes(y = CRABP2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime RA Synthesis Genes")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  #stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'grey40', se = FALSE) +
  stat_smooth(aes(y = TGFA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue', se = FALSE) +
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red4', se = FALSE) +
  stat_smooth(aes(y = BMP2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod1', se = FALSE) +
  stat_smooth(aes(y = BMP7), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod3', se = FALSE) +  
  stat_smooth(aes(y = SMAD7), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod', se = FALSE) +
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid', se = FALSE) +
  theme_classic() +
  ggtitle("Glover2023 Regulators of Fingerprint Ridge")
```

```{r}
plot_cells(cds_glover,
           color_cells_by = "pseudotime",
           group_cells_by = "cluster",
           label_cell_groups = FALSE,
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           label_roots = TRUE,
           cell_size = 0.75,
           trajectory_graph_color = "grey40", show_trajectory_graph = TRUE)#visualize pseudotime umap and display root node
```
