---
title: "Sean_Liu2022_Mouse_P46_Reanalysis"
output: html_notebook
---

#Load packages
```{r}
library(dplyr)
library(Seurat)
library(tictoc)
library(reticulate)
library(limma)#from Biocmanager
library(ggplot2)
library(viridisLite)
library(patchwork)
library(future)
library(RColorBrewer)
divergentcolors_RYB <- colorRampPalette(rev(brewer.pal(11, "RdYlBu")))#take our 3-color gradient (-, 0, +) and make a gradient of colors
plan('multisession', workers = 5)#cores = 5
options(future.globals.maxSize = 100000 * 1024^2)
```


#Load data
```{r}
# Load the datasets
WT1.data <- Read10X("~/Documents/Liu2022/P46_WT1")
WT2.data <- Read10X("~/Documents/Liu2022/P46_WT2")

# Initialize the Seurat object with the raw (non-normalized data).
Liu2022_1 <- CreateSeuratObject(counts = WT1.data, project = "WT1", min.cells = 3, min.features = 200)
Liu2022_2 <- CreateSeuratObject(counts = WT2.data, project = "WT2", min.cells = 3, min.features = 200)

Liu2022_1
Liu2022_2

P46_Mouse <- merge(x = Liu2022_1, y = c(Liu2022_2), add.cell.ids = c('P46_1', 'P46_2'), project = 'Liu2022_Mouse_P46_scRNA')
```

#QC
```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
P46_Mouse[["percent.mt"]] <- PercentageFeatureSet(P46_Mouse, pattern = "^mt-")

# Visualize QC metrics as a violin plot
VlnPlot(P46_Mouse, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(P46_Mouse, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(P46_Mouse, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```

#QC Filter
```{r}
P46_Mouse
P46_Mouse <- subset(x = P46_Mouse, 
                          subset = percent.mt < 10)
P46_Mouse

# Visualize QC metrics as a violin plot
VlnPlot(P46_Mouse, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

## SCTransform (Normalize, Scale Data)
```{r, warning=FALSE, message=FALSE}
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
P46_Mouse <- SCTransform(P46_Mouse, vars.to.regress = "percent.mt", verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
P46_Mouse <- RunPCA(P46_Mouse, verbose = FALSE)
ElbowPlot(P46_Mouse, ndims = 50)
```

## UMAP (initial)
Pick dimensions to use based on ElbowPlot above -- try to pick dimensions w/ > 1 standard deviation. In this case, up to 40 dimensions is greater than 5 standard deviations before it levels off (suggesting decline in variability).
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
tic('Running UMAP & Making DimPlots')
#run the UMAP function using dimensions informed by elbow plot
P46_Mouse <- RunUMAP(P46_Mouse, dims = 1:40, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
P46_Mouse <- FindNeighbors(P46_Mouse, dims = 1:40, verbose = FALSE)
P46_Mouse <- FindClusters(P46_Mouse, verbose = FALSE, algorithm = 3)#default is algorithm = 1/2 (Louvain), 3 = SLM (scATAC), 4 = Leiden (not installed on server currently - need to pip install leidenalg later)
DimPlot(P46_Mouse, label = TRUE, pt.size = 0.5, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P46_Mouse, label = FALSE, pt.size = 0.5) + NoLegend()#no labels (to label in photoshop)
DimPlot(P46_Mouse, group.by = 'orig.ident', pt.size = 0.5)
#DimPlot(P46_Mouse, label = FALSE)#no numbers on clusters, legend off to side
toc()

#Check QC Metrics on Clustering
FeaturePlot(P46_Mouse, features = c('nFeature_RNA', 'nCount_RNA'))
FeaturePlot(P46_Mouse, features = c('percent.mt'))
```

#FeaturePlot
Note: Cd207 encodes Langerin (marker of epidermal Langerhans Cells)
```{r fig.height=7, fig.width=7, message=FALSE, warning=FALSE}
## FeaturePlots: view single-cell gene expression across clusters
GOI <- c('Pdgfra', 'Lum', 'Runx1t1', 'Lef1', 'Sox2', 'Crabp1', 'Prdm1', 'Runx3', 'Alx4', 'Prlr', 'a',
         'Mfap5', 'Dpp4', 'Col3a1', 'Pparg', 'Fabp4', 
         'Pdgfrb', 'Vim', 'Lrrc17', 'Rgs5', 'Acta2', 'Lrrc15', 'Itga8', 'Itga6', 'Itgb4',
         'Krt14', 'Krt15', 'Sox9', 'Krt10', 'Krt1', 'Cnfn', 'Krt23', 'Krt8', 'Krt18', 'Krt19',
         'Mki67', 'Coch', 'Alas2', 'Hba-a1', 'Ptprc', 'Cd86', 'Cd3g', 'Cd207', 'Msc',
         'Vegfa', 'Pecam1', 'Cdh5', 'Flt1', 'Kdr', 'Flt4', 'Lyve1', 'Sox10', 'Mbp', 'Dct', 'Tyrp1', 'Dlk2', 'Dll1', 'Angptl1', 'Angptl2')

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = P46_Mouse,
                    features = GOI[n_plot],
                    pt.size = 0.5,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()
```


### Change Resolution
```{r message=FALSE, warning=FALSE}
P46_Mouse <- FindClusters(P46_Mouse, verbose = FALSE, algorithm = 3, resolution = 0.8)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(P46_Mouse, label = TRUE, pt.size = 0.5, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P46_Mouse, label = FALSE, pt.size = 0.5) + NoLegend()#no labels (to label in photoshop)
DimPlot(P46_Mouse, label = FALSE, pt.size = 0.5, group.by = 'orig.ident')#no labels (to label in photoshop)
```

## Label the Clusters by Cell Type
```{r message=FALSE, warning=FALSE}
#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids.all <- c('0 Basal Krtno', '1 Diff Krtno', '2 Basal Krtno', '3 Basal Krtno', '4 Immune Cell',
                     '5 Immune Cell', '6 Fibro', '7 Immune Cell', '8 DP', '9 Fibro', 
                     '10 LBV', '11 Basal Krtno', '12 Schwann', '13 Immune Cell', '14 DP', 
                     '15 Erythro', '16 Diff Krtno', '17 Div Krtno', '18 Fibro', '19 Schwann',
                     '20 Pericyte', '21 Fascia', '22 Immune Cell', '23 Diff Krtno', '24 Fibro', 
                     '25 BV', '26 Krtno', '27 Melanocyte', '28 Immune Cell', '29 Immune Cell')
names(new.cluster.ids.all) <- levels(P46_Mouse)
P46_Mouse <- RenameIdents(P46_Mouse, new.cluster.ids.all)
```

Save the R object
```{r}
save(P46_Mouse, file = '~/Desktop/Sean_CellChat/Object/Liu2022_P46_Mouse_AllClusters.RData')#final object w/ labels
```

## Recolor the UMAP clusters by cell lineage
```{r fig.height=7, fig.width=7}
path_out <- '~/Desktop/Sean_CellChat/UMAP/'
p1 <- DimPlot(P46_Mouse, label = FALSE, pt.size = 1.0, 
              cols = c('#16539e', '#4d8ad5', '#0c3695', '#002783', '#cc0000',
                     '#BB0000', '#238b45', '#900101', '#006837', '#52ba55', 
                     '#88419d', '#00206a', '#fc8d59', '#790001', '#00441b', 
                     '#DD0000', '#64a7e9', '#2565b5', '#41ab5d', '#ffab4b',
                     '#7fcdbb', '#a1d99b', '#AD0000', '#4d91d5', '#78c679', 
                     '#8c6bb1', '#3b7bca', '#f16913', '#DD0000', '#640001')) + NoLegend()#manual colors w/o legend
p1
ggsave(plot = p1, filename = 'Liu2022 Mouse P46 UMAP (All Clusters).png', path = path_out,
         width = 7, height = 7, units = 'in')
```
