---
title: "R Notebook"
output: html_notebook
---

#0a. Load packages
Updated to Seurat version 4.3.0 (from R CRAN) from 4.1.1 on 6/19/23 b/c SeuratWrappers requires >= Seurat version 4.2.0
```{r}
#install.packages('textshaping')#needed ~"sudo apt install libharfbuzz-dev libfribidi-dev" first
#install.packages('Cairo)#needed ~"sudo apt install libxt-dev", ~"sudo apt-get install libcairo2-dev" first
#install.packages('units')#needed ~"sudo apt-get install libudunits2-dev" first
#install.packages('ragg')#needed ~"libfreetype6-dev libpng-dev libtiff5-dev libjpeg-dev" first
#install.packages('terra')#needed ~"sudo apt-get install gdal-bin libgdal-dev" first
#install.packages('sf')#
#install.packages('spdep')#
#install.packages('ggrastr')#

library(monocle3)#installed via devtools::install_github('cole-trapnell-lab/monocle3')
library(Seurat)
library(SeuratWrappers)#installed via remotes::install_github('satijalab/seurat-wrappers')
library(patchwork)
library(ggplot2)
#library(dplyr)
set.seed(1234)
#library(slingshot)#installed BiocManager::install("kstreet13/slingshot"), alternative to monocle3, didn't use
```

#1. E90 Pig Keratinocytes
##Convert Seurat object (Seurat) to Cell Data Set object (Monocle3) and Run Pseudotime
Adapted from a vignette for performing pseudotime analysis and other posts related to converting a Seurat object to work in Monocle3: https://cole-trapnell-lab.github.io/monocle3/docs/trajectories/, https://github.com/satijalab/seurat-wrappers/issues/54, https://ucdavis-bioinformatics-training.github.io/2021-August-Advanced-Topics-in-Single-Cell-RNA-Seq-Trajectory-and-Velocity/data_analysis/monocle_fixed, https://github.com/satijalab/seurat/issues/1658, https://github.com/cole-trapnell-lab/monocle-release/issues/388, https://www.biostars.org/p/9540648/
```{r}
#plot <- DimPlot(E90_krtno)

#selectedcells <- CellSelector(plot = plot)#grab the cells in the umap to keep
#selectedcells
# Automatically set the identity class of selected cells and return a new Seurat object
#E90_krtno_sub <- CellSelector(plot = plot, object = E90_krtno, ident = 'E90SelectedCells')
#DimPlot(E90_krtno_sub)

cds_e90sub <- as.cell_data_set(E90_krtno, group.by = 'SCT_snn_res.0.5')#function from SeuratWrappers to convert to Monocle3 object, graph argument specifies the clustering to transfer (assay to use, and active resolution) to @clusters
#cds_e90sub <- cluster_cells(cds_e90sub, reduction_method = "UMAP", group_cells_by = 'seurat_clusters')#

#Add some info not carried over by SeuratWrappers
cds_e90sub <- estimate_size_factors(cds_e90sub)#add size factor metadata to CDS object
cds_e90sub@rowRanges@elementMetadata@listData[["gene_short_name"]] <- rownames(E90_krtno)#correct gene names in CDS object

#Validate UMAP coordinate transfer from Seurat object -> CDS object
temp1 <- E90_krtno@reductions$umap@cell.embeddings#large matrix with rows cell barcodes and 2 columns for the Seurat x/y UMAP coordinates
temp2 <- cds_e90sub@int_colData@listData$reducedDims$UMAP#matrix with rows cell barcodes and 2 columns for x/y UMAP coordinates
table(temp1 == temp2)#if all TRUE, seuratwrapper successfully transferred umap coordinates to CDS object via the graph argument
#cds_e90sub@int_colData@listData$reducedDims$UMAP <- E90_krtno@reductions$umap@cell.embeddings#not needed due to graph argument when making CDS object

#need to transfer Seurat PCA since this wasn't transferred by seurat wrapper
reducedDim(cds_e90sub, type = "PCA") <- E90_krtno@reductions$pca@cell.embeddings#transfer the matrix results of Seurat PCA to CDS
#cds_e90sub@preprocess_aux$prop_var_expl <- E90_krtno@reductions$pca@stdev#transfer numerical vector containing stdev values per PCA dimension

#partitions and cleaner subset still need to be created for learn_graph()
sub_cells <- choose_cells(cds_e90sub, return_list = TRUE)#get cell barcodes to make new seurat subset
plot <- DimPlot(E90_krtno, cells = as.vector(sub_cells))
plot
E90_krtno_sub <- subset(E90_krtno, cells = sub_cells)#subset to the clean clusters
```

Based on ElbowPlot, pick major PCs for next steps
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
E90_krtno_sub <- RunUMAP(E90_krtno_sub, dims = 1:30, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
E90_krtno_sub <- FindNeighbors(E90_krtno_sub, dims = 1:30, verbose = FALSE)
E90_krtno_sub <- FindClusters(E90_krtno_sub, verbose = FALSE, algorithm = 3, resolution = 0.6)#default is algorithm = 1 (Louvain), 3 = SLM
toc()
DimPlot(E90_krtno_sub, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(E90_krtno_sub, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

### View Gene Expression of Keratinocyte Heterogeneity Markers
```{r fig.height=5, fig.width=5}
## FeaturePlots: view single-cell gene expression across clusters
DimPlot(E90_krtno_sub, label = TRUE, pt.size = 2, label.size = 6.0)

GOI <- c('KRT14', 'KRT5', 'KRT10', 'KRT18', 'KRT8', 'CALML5', 'CDSN', 'CNFN', 'CDH1',
         'TOP2A', 'ITGA6', 'ITGB1', 'SOX6', 'SOX9', 'COL17A1', 'DLL1', 'KRT15', 'VEGFA', 'KRT10', 'nFeature_SCT', 'LEF1', 'CTNNB1')

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = E90_krtno_sub,
                    features = GOI[n_plot],
                    pt.size = 1.0,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()

```

### Change Resolution
```{r message=FALSE, warning=FALSE}
E90_krtno_sub <- FindClusters(E90_krtno_sub, verbose = FALSE, algorithm = 3, resolution = 0.4)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(E90_krtno_sub, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(E90_krtno_sub, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

Save Seurat object
```{r}
save(E90_krtno_sub, file = '/t2tb/Object/E90_pig_krtno_sub.RData')
```

##Fig Quality DimPlot
```{r fig.height=5, fig.width=7}
path_out <- '~/Desktop/Sean_CellChat/UMAP/'
p1<- DimPlot(E90_krtno_sub, label = FALSE, pt.size = 1.0,
        cols = c('#0c3695', '#f0b627', '#107de9', '#225ea8', '#ff692b', '#00206a', '#6b92c3')) + NoLegend()#blank
p1

ggsave(plot = p1, filename = 'Pig E90 UMAP (Krtno Sub).png', path = path_out,
         width = 7, height = 7, units = 'in')
```

##Pseudotime
```{r}
cds_e90sub <- as.cell_data_set(E90_krtno_sub, group.by = 'SCT_snn_res.0.4')#function from SeuratWrappers to convert to Monocle3 object, graph argument specifies the clustering to transfer (assay to use, and active resolution) to @clusters
#cds_e90sub <- cluster_cells(cds_e90sub, reduction_method = "UMAP", group_cells_by = 'seurat_clusters')#

#Add some info not carried over by SeuratWrappers
cds_e90sub <- estimate_size_factors(cds_e90sub)#add size factor metadata to CDS object
cds_e90sub@rowRanges@elementMetadata@listData[["gene_short_name"]] <- rownames(E90_krtno_sub)#correct gene names in CDS object

#Validate UMAP coordinate transfer from Seurat object -> CDS object
temp1 <- E90_krtno_sub@reductions$umap@cell.embeddings#large matrix with rows cell barcodes and 2 columns for the Seurat x/y UMAP coordinates
temp2 <- cds_e90sub@int_colData@listData$reducedDims$UMAP#matrix with rows cell barcodes and 2 columns for x/y UMAP coordinates
table(temp1 == temp2)#if all TRUE, seuratwrapper successfully transferred umap coordinates to CDS object via the graph argument
#cds_e90sub@int_colData@listData$reducedDims$UMAP <- E90_krtno_sub@reductions$umap@cell.embeddings#not needed due to graph argument when making CDS object

#need to transfer Seurat PCA since this wasn't transferred by seurat wrapper
reducedDim(cds_e90sub, type = "PCA") <- E90_krtno_sub@reductions$pca@cell.embeddings#transfer the matrix results of Seurat PCA to CDS
#cds_e90sub@preprocess_aux$prop_var_expl <- E90_krtno_sub@reductions$pca@stdev#transfer numerical vector containing stdev values per PCA dimension


manual_partition <- as.vector(E90_krtno_sub@meta.data$seurat_clusters)#get cluster id #s to construct manual partition from a factor
names(manual_partition) <- rownames(E90_krtno_sub@meta.data)#
manual_partition <- as.factor(manual_partition)
head(manual_partition)#view ordered levels
levels(manual_partition) <- c(1, 1, 1, 1, 1, 1, 1)#replace cluster id levels with new ids for partitioning in Monocle3
table(levels(manual_partition))#validate replacement
cds_e90sub@clusters@listData$UMAP$partitions <- manual_partition

#do pseudotime
cds_e90sub <- learn_graph(cds_e90sub, use_partition = TRUE, verbose = TRUE)

p2 <- plot_cells(cds_e90sub, reduction_method = 'UMAP',
                 color_cells_by = 'seurat_clusters', 
                 group_label_size = 5, cell_size = 1,
                 show_trajectory_graph = TRUE)#is a ggplot2 object, so can custom color same as with Seurat's DimPlot
p2

cds_e90sub <- order_cells(cds_e90sub)#choose root node
plot_cells(cds_e90sub,
           color_cells_by = "pseudotime",
           group_cells_by = "cluster",
           label_cell_groups = FALSE,
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           label_roots = FALSE,
           cell_size = 0.75,
           trajectory_graph_color = "grey40")
plot_cells(cds_e90sub,
           color_cells_by = "pseudotime",
           group_cells_by = "cluster",
           label_cell_groups = FALSE,
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           label_roots = FALSE,
           cell_size = .75,
           show_trajectory_graph = FALSE)


```

Save CDS object
```{r}
save(cds_e90sub, file = '/t2tb/Object/E90_pig_krtno_sub_CDS.RData')
```

```{r}
pseudotime_df <- as.data.frame(t(cds_e90sub@assays@data@listData$logcounts))#get cell by gene expression matrix
pseudotime_df$Barcode <- rownames(pseudotime_df)#get column of barcode values
pseudotime_df$Pseudotime <- pseudotime(cds_e90sub)#get pseudotime values

temp_df <- data.frame("Barcode" = rownames(as.data.frame(cds_e90sub@clusters@listData$UMAP$clusters)), 
                      "Cluster" = cds_e90sub@clusters@listData$UMAP$clusters)#get cluster ids

pseudotime_df <- merge(temp_df, pseudotime_df)#filter to shared rows in subset cds
pseudotime_df <- pseudotime_df[pseudotime_df$Pseudotime != Inf, ]#remove rows w/ inf pseudotime value

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey40', se = FALSE) +
  stat_smooth(aes(y = PTCH1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue', se = FALSE) +
  stat_smooth(aes(y = PTCH2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +  
  stat_smooth(aes(y = SHH), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +   
  stat_smooth(aes(y = WNT10B), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red3', se = FALSE) +
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red4', se = FALSE) +
  stat_smooth(aes(y = EDA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid1', se = FALSE) +   
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal Bud DEGs")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey40', se = FALSE) +
  stat_smooth(aes(y = SHH), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +   
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red4', se = FALSE) +
  #stat_smooth(aes(y = EDA), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'orchid1', se = FALSE) +   
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal Bud")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey60', se = FALSE, linetype = 'dashed') +
  stat_smooth(aes(y = SHH), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'blue', se = FALSE) +   
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red', se = FALSE) +
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'green', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal Bud (Figure)")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  #stat_smooth(aes(y = KRT15), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +
  stat_smooth(aes(y = KRT14), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue3', se = FALSE) +  
  stat_smooth(aes(y = KRT5), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +
  stat_smooth(aes(y = KRT10), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod2', se = FALSE) +
  stat_smooth(aes(y = KRT1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkorange', se = FALSE) +
  theme_classic() + ylim(0, 6) +
  ggtitle("Pseudotime Keratins")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey30', se = FALSE) +
  stat_smooth(aes(y = ITGA2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey50', se = FALSE) +
  stat_smooth(aes(y = DLL1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +
  stat_smooth(aes(y = SOX6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'peru', se = FALSE) +  
  stat_smooth(aes(y = DLK2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue3', se = FALSE) +
  stat_smooth(aes(y = NRG1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +
  stat_smooth(aes(y = PHLDB2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  stat_smooth(aes(y = LAMB3), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid4', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal IFE DEGs")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = VEGFA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid1', se = FALSE) +
    stat_smooth(aes(y = VEGFC), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  stat_smooth(aes(y = ANGPTL1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple1', se = FALSE) +
  stat_smooth(aes(y = ANGPTL2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple3', se = FALSE) +
  stat_smooth(aes(y = ANGPTL4), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple4', se = FALSE) +
  stat_smooth(aes(y = PDGFA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgreen', se = FALSE) +
  stat_smooth(aes(y = PDGFB), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'lightseagreen', se = FALSE) +  
  stat_smooth(aes(y = PDGFC), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'green3', se = FALSE) +
  #stat_smooth(aes(y = COCH), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'darkgreen', se = FALSE) +
  #stat_smooth(aes(y = COL2A1), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'lightseagreen', se = FALSE) +  
  theme_classic() +
  ggtitle("Pseudotime Dermal Pocket Genes")


ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = RDH10), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid1', se = FALSE) +
  stat_smooth(aes(y = DHRS3), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkred', se = FALSE) +
  stat_smooth(aes(y = ALDH1A1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple3', se = FALSE) +
  stat_smooth(aes(y = ALDH1A3), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple4', se = FALSE) +
  stat_smooth(aes(y = RARA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +
  stat_smooth(aes(y = RARG), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue3', se = FALSE) +
  stat_smooth(aes(y = CRABP2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime RA Synthesis Genes")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  #stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'grey40', se = FALSE) +
  stat_smooth(aes(y = TGFA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue', se = FALSE) +
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red4', se = FALSE) +
  stat_smooth(aes(y = BMP2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod1', se = FALSE) +
  stat_smooth(aes(y = BMP7), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod2', se = FALSE) +  
  stat_smooth(aes(y = SMAD7), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod', se = FALSE) +
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid', se = FALSE) +
  theme_classic() +
  ggtitle("Glover2023 Regulators of Fingerprint Ridge")
```

```{r}
plot_cells(cds_e90sub,
           color_cells_by = "pseudotime",
           group_cells_by = "cluster",
           label_cell_groups = FALSE,
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           label_roots = TRUE,
           cell_size = 0.75,
           trajectory_graph_color = "grey40", show_trajectory_graph = TRUE)#visualize pseudotime umap and display root node
```


#2. P10 Pig Keratinocytes
```{r}
cds_p10sub <- as.cell_data_set(P10_krtno, group.by = 'SCT_snn_res.0.15')#function from SeuratWrappers to convert to Monocle3 object, graph argument specifies the clustering to transfer (assay to use, and active resolution) to @clusters
#cds_p10sub <- cluster_cells(cds_p10sub, reduction_method = "UMAP", group_cells_by = 'seurat_clusters')#

#Add some info not carried over by SeuratWrappers
cds_p10sub <- estimate_size_factors(cds_p10sub)#add size factor metadata to CDS object
cds_p10sub@rowRanges@elementMetadata@listData[["gene_short_name"]] <- rownames(P10_krtno)#correct gene names in CDS object

#Validate UMAP coordinate transfer from Seurat object -> CDS object
temp1 <- P10_krtno@reductions$umap@cell.embeddings#large matrix with rows cell barcodes and 2 columns for the Seurat x/y UMAP coordinates
temp2 <- cds_p10sub@int_colData@listData$reducedDims$UMAP#matrix with rows cell barcodes and 2 columns for x/y UMAP coordinates
table(temp1 == temp2)#if all TRUE, seuratwrapper successfully transferred umap coordinates to CDS object via the graph argument
#cds_p10sub@int_colData@listData$reducedDims$UMAP <- P10_krtno@reductions$umap@cell.embeddings#not needed due to graph argument when making CDS object

#need to transfer Seurat PCA since this wasn't transferred by seurat wrapper
reducedDim(cds_p10sub, type = "PCA") <- P10_krtno@reductions$pca@cell.embeddings#transfer the matrix results of Seurat PCA to CDS
#cds_p10sub@preprocess_aux$prop_var_expl <- P10_krtno@reductions$pca@stdev#transfer numerical vector containing stdev values per PCA dimension

#partitions and cleaner subset still need to be created for learn_graph()
sub_cells <- choose_cells(cds_p10sub, return_list = TRUE)#get cell barcodes to make new seurat subset
plot <- DimPlot(P10_krtno, cells = as.vector(sub_cells))
plot
P10_krtno_sub <- subset(P10_krtno, cells = sub_cells)#subset to the clean clusters
```

Based on ElbowPlot, pick major PCs for next steps
#new UMAP
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
P10_krtno_sub <- RunUMAP(P10_krtno_sub, dims = 1:20, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
P10_krtno_sub <- FindNeighbors(P10_krtno_sub, dims = 1:20, verbose = FALSE)
P10_krtno_sub <- FindClusters(P10_krtno_sub, verbose = FALSE, algorithm = 3, resolution = 0.6)#default is algorithm = 1 (Louvain), 3 = SLM
toc()
DimPlot(P10_krtno_sub, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P10_krtno_sub, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

### View Gene Expression of Keratinocyte Heterogeneity Markers
```{r fig.height=5, fig.width=5}
## FeaturePlots: view single-cell gene expression across clusters
DimPlot(P10_krtno_sub, label = TRUE, pt.size = 2, label.size = 6.0)

GOI <- c('KRT14', 'KRT5', 'KRT10', 'KRT18', 'KRT8', 'CALML5', 'CDSN', 'CNFN', 'CDH1',
         'TOP2A', 'ITGA6', 'ITGB1', 'SOX6', 'SOX9', 'COL17A1', 'DLL1', 'KRT15', 'EREG', 'nFeature_SCT', 'LEF1', 'CTNNB1')

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = P10_krtno_sub,
                    features = GOI[n_plot],
                    pt.size = 1.5,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()

```

### Change Resolution
```{r message=FALSE, warning=FALSE}
P10_krtno_sub <- FindClusters(P10_krtno_sub, verbose = FALSE, algorithm = 3, resolution = 0.3)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(P10_krtno_sub, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P10_krtno_sub, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

Save Seurat object
```{r}
save(P10_krtno_sub, file = '/t2tb/Object/P10_pig_krtno_sub.RData')
```

##Fig Quality DimPlot
```{r fig.height=5, fig.width=7}
path_out <- '~/Desktop/Sean_CellChat/UMAP/'
p1<- DimPlot(P10_krtno_sub, label = FALSE, pt.size = 1.5,
        cols = c('#f09527', '#ffab4b', '#ff692b', '#107de9', '#00206a', 
                 '#225ea8', '#f0b627', '#0c3695', '#d04800')) + NoLegend()#blank
p1

ggsave(plot = p1, filename = 'Pig P10 UMAP (Krtno Sub).png', path = path_out,
         width = 7, height = 7, units = 'in')
```


##Pseudotime
```{r}
cds_p10sub <- as.cell_data_set(P10_krtno_sub, group.by = 'SCT_snn_res.0.3')#function from SeuratWrappers to convert to Monocle3 object, graph argument specifies the clustering to transfer (assay to use, and active resolution) to @clusters
#cds_p10sub <- cluster_cells(cds_p10sub, reduction_method = "UMAP", group_cells_by = 'seurat_clusters')#

#Add some info not carried over by SeuratWrappers
cds_p10sub <- estimate_size_factors(cds_p10sub)#add size factor metadata to CDS object
cds_p10sub@rowRanges@elementMetadata@listData[["gene_short_name"]] <- rownames(P10_krtno_sub)#correct gene names in CDS object

#Validate UMAP coordinate transfer from Seurat object -> CDS object
temp1 <- P10_krtno_sub@reductions$umap@cell.embeddings#large matrix with rows cell barcodes and 2 columns for the Seurat x/y UMAP coordinates
temp2 <- cds_p10sub@int_colData@listData$reducedDims$UMAP#matrix with rows cell barcodes and 2 columns for x/y UMAP coordinates
table(temp1 == temp2)#if all TRUE, seuratwrapper successfully transferred umap coordinates to CDS object via the graph argument
#cds_p10sub@int_colData@listData$reducedDims$UMAP <- P10_krtno_sub@reductions$umap@cell.embeddings#not needed due to graph argument when making CDS object

#need to transfer Seurat PCA since this wasn't transferred by seurat wrapper
reducedDim(cds_p10sub, type = "PCA") <- P10_krtno_sub@reductions$pca@cell.embeddings#transfer the matrix results of Seurat PCA to CDS
#cds_p10sub@preprocess_aux$prop_var_expl <- P10_krtno_sub@reductions$pca@stdev#transfer numerical vector containing stdev values per PCA dimension


manual_partition <- as.vector(P10_krtno_sub@meta.data$seurat_clusters)#get cluster id #s to construct manual partition from a factor
names(manual_partition) <- rownames(P10_krtno_sub@meta.data)#
manual_partition <- as.factor(manual_partition)
head(manual_partition)#view ordered levels
levels(manual_partition) <- c(1, 1, 1, 1, 1, 1, 1, 1, 1)#replace cluster id levels with new ids for partitioning in Monocle3
table(levels(manual_partition))#validate replacement
cds_p10sub@clusters@listData$UMAP$partitions <- manual_partition

#do pseudotime
cds_p10sub <- learn_graph(cds_p10sub, use_partition = TRUE, verbose = TRUE)

p2 <- plot_cells(cds_p10sub, reduction_method = 'UMAP',
                 color_cells_by = 'seurat_clusters', 
                 group_label_size = 5, cell_size = 1,
                 show_trajectory_graph = TRUE)#is a ggplot2 object, so can custom color same as with Seurat's DimPlot
p2

cds_p10sub <- order_cells(cds_p10sub)#choose root node
plot_cells(cds_p10sub,
           color_cells_by = "pseudotime",
           group_cells_by = "cluster",
           label_cell_groups = FALSE,
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           label_roots = FALSE,
           cell_size = 0.75,
           trajectory_graph_color = "grey60")
plot_cells(cds_p10sub,
           color_cells_by = "pseudotime",
           group_cells_by = "cluster",
           label_cell_groups = FALSE,
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           label_roots = FALSE,
           cell_size = .75,
           show_trajectory_graph = FALSE)


```

Save CDS object
```{r}
save(cds_p10sub, file = '/t2tb/Object/P10_pig_krtno_sub_CDS.RData')
```

```{r}
pseudotime_df <- as.data.frame(t(cds_p10sub@assays@data@listData$logcounts))#get cell by gene expression matrix
pseudotime_df$Barcode <- rownames(pseudotime_df)#get column of barcode values
pseudotime_df$Pseudotime <- pseudotime(cds_p10sub)#get pseudotime values

temp_df <- data.frame("Barcode" = rownames(as.data.frame(cds_p10sub@clusters@listData$UMAP$clusters)), 
                      "Cluster" = cds_p10sub@clusters@listData$UMAP$clusters)#get cluster ids

pseudotime_df <- merge(temp_df, pseudotime_df)#filter to shared rows in subset cds
pseudotime_df <- pseudotime_df[pseudotime_df$Pseudotime != Inf, ]#remove rows w/ inf pseudotime value

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey40', se = FALSE) +
  stat_smooth(aes(y = PTCH1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue', se = FALSE) +
  stat_smooth(aes(y = PTCH2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +  
  #stat_smooth(aes(y = SHH), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +   
  stat_smooth(aes(y = WNT10B), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red3', se = FALSE) +
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red4', se = FALSE) +
  stat_smooth(aes(y = EDA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid1', se = FALSE) +   
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal Bud DEGs")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey40', se = FALSE) +
  #stat_smooth(aes(y = SHH), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +   
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red4', se = FALSE) +
  #stat_smooth(aes(y = EDA), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'orchid1', se = FALSE) +   
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal Bud")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey60', se = FALSE, linetype = 'dashed') +
  #stat_smooth(aes(y = SHH), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'blue', se = FALSE) +   
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red', se = FALSE) +
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'green', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal Bud (Figure)")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  #stat_smooth(aes(y = KRT15), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +
  stat_smooth(aes(y = KRT14), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue3', se = FALSE) +  
  stat_smooth(aes(y = KRT5), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +
  stat_smooth(aes(y = KRT10), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod2', se = FALSE) +
  stat_smooth(aes(y = KRT1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkorange', se = FALSE) +
  theme_classic() + ylim(0, 6) +
  ggtitle("Pseudotime Keratins")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey30', se = FALSE) +
  stat_smooth(aes(y = ITGA2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'grey50', se = FALSE) +
  stat_smooth(aes(y = DLL1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +
  stat_smooth(aes(y = SOX6), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'peru', se = FALSE) +  
  stat_smooth(aes(y = DLK2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue3', se = FALSE) +
  stat_smooth(aes(y = NRG1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +
  stat_smooth(aes(y = PHLDB2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  stat_smooth(aes(y = LAMB3), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid4', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime Basal IFE DEGs")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = VEGFA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid1', se = FALSE) +
    stat_smooth(aes(y = VEGFC), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid3', se = FALSE) +
  stat_smooth(aes(y = ANGPTL1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple1', se = FALSE) +
  stat_smooth(aes(y = ANGPTL2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple3', se = FALSE) +
  stat_smooth(aes(y = ANGPTL4), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple4', se = FALSE) +
  stat_smooth(aes(y = PDGFA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgreen', se = FALSE) +
  stat_smooth(aes(y = PDGFB), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'lightseagreen', se = FALSE) +  
  stat_smooth(aes(y = PDGFC), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'green3', se = FALSE) +
  #stat_smooth(aes(y = COCH), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'darkgreen', se = FALSE) +
  #stat_smooth(aes(y = COL2A1), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'lightseagreen', se = FALSE) +  
  theme_classic() +
  ggtitle("Pseudotime Dermal Pocket Genes")


ggplot(pseudotime_df, aes(x = Pseudotime)) +
  stat_smooth(aes(y = RDH10), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid1', se = FALSE) +
  stat_smooth(aes(y = DHRS3), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkred', se = FALSE) +
  stat_smooth(aes(y = ALDH1A1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple3', se = FALSE) +
  stat_smooth(aes(y = ALDH1A3), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'mediumpurple4', se = FALSE) +
  stat_smooth(aes(y = RARA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue4', se = FALSE) +
  stat_smooth(aes(y = RARG), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue3', se = FALSE) +
  stat_smooth(aes(y = CRABP2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue1', se = FALSE) +
  theme_classic() +
  ggtitle("Pseudotime RA Synthesis Genes")

ggplot(pseudotime_df, aes(x = Pseudotime)) +
  #stat_smooth(aes(y = ITGA6), method = "gam", formula = y ~ s(x), 
  #            size = 1, alpha = .75, color = 'grey40', se = FALSE) +
  stat_smooth(aes(y = TGFA), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'steelblue', se = FALSE) +
  stat_smooth(aes(y = LEF1), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'red4', se = FALSE) +
  stat_smooth(aes(y = BMP2), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod1', se = FALSE) +
  stat_smooth(aes(y = BMP7), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod2', se = FALSE) +  
  stat_smooth(aes(y = SMAD7), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'darkgoldenrod', se = FALSE) +
  stat_smooth(aes(y = EDAR), method = "gam", formula = y ~ s(x), 
              size = 1, alpha = .75, color = 'orchid', se = FALSE) +
  theme_classic() +
  ggtitle("Glover2023 Regulators of Fingerprint Ridge")
```

```{r}
plot_cells(cds_p10sub,
           color_cells_by = "pseudotime",
           group_cells_by = "cluster",
           label_cell_groups = FALSE,
           label_groups_by_cluster=FALSE,
           label_leaves=FALSE,
           label_branch_points=FALSE,
           label_roots = TRUE,
           cell_size = 0.75,
           trajectory_graph_color = "grey40", show_trajectory_graph = TRUE)#visualize pseudotime umap and display root node
```


